// <auto-generated/>
// Contents of: hl7.fhir.r4.core version: 4.0.1
  // Primitive Naming Style: None
  // Complex Type / Resource Naming Style: PascalCase
  // Interaction Naming Style: None
  // Extension Support: NonPrimitive
// Minimum TypeScript Version: 3.7
import * as fhir from '../fhir'
/**
 * A sequence that is used as a reference to describe variants that are present in a sequence analyzed.
 */
export type IMolecularSequenceReferenceSeq = fhir.IBackboneElement & {
  /**
   * Structural unit composed of a nucleic acid molecule which controls its own replication through the interaction of specific proteins at one or more origins of replication ([SO:0000340](http://www.sequenceontology.org/browser/current_svn/term/SO:0000340)).
   */
  chromosome?: fhir.ICodeableConcept|undefined;
  /**
   * The Genome Build used for reference, following GRCh build versions e.g. 'GRCh 37'.  Version number must be included if a versioned release of a primary build was used.
   */
  genomeBuild?: string|undefined;
  _genomeBuild?: fhir.IFhirElement|undefined;
  /**
   * A relative reference to a DNA strand based on gene orientation. The strand that contains the open reading frame of the gene is the "sense" strand, and the opposite complementary strand is the "antisense" strand.
   */
  orientation?: MolecularSequenceReferenceSeqOrientationEnum|undefined;
  _orientation?: fhir.IFhirElement|undefined;
  /**
   * Reference identifier of reference sequence submitted to NCBI. It must match the type in the MolecularSequence.type field. For example, the prefix, “NG_” identifies reference sequence for genes, “NM_” for messenger RNA transcripts, and “NP_” for amino acid sequences.
   */
  referenceSeqId?: fhir.ICodeableConcept|undefined;
  /**
   * A pointer to another MolecularSequence entity as reference sequence.
   */
  referenceSeqPointer?: fhir.IReference|undefined;
  /**
   * A string like "ACGT".
   */
  referenceSeqString?: string|undefined;
  _referenceSeqString?: fhir.IFhirElement|undefined;
  /**
   * An absolute reference to a strand. The Watson strand is the strand whose 5'-end is on the short arm of the chromosome, and the Crick strand as the one whose 5'-end is on the long arm.
   */
  strand?: MolecularSequenceReferenceSeqStrandEnum|undefined;
  _strand?: fhir.IFhirElement|undefined;
  /**
   * End position of the window on the reference sequence. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  windowEnd?: number|undefined;
  _windowEnd?: fhir.IFhirElement|undefined;
  /**
   * Start position of the window on the reference sequence. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  windowStart?: number|undefined;
  _windowStart?: fhir.IFhirElement|undefined;
}
/**
 * The definition of variant here originates from Sequence ontology ([variant_of](http://www.sequenceontology.org/browser/current_svn/term/variant_of)). This element can represent amino acid or nucleic sequence change(including insertion,deletion,SNP,etc.)  It can represent some complex mutation or segment variation with the assist of CIGAR string.
 */
export type IMolecularSequenceVariant = fhir.IBackboneElement & {
  /**
   * Extended CIGAR string for aligning the sequence with reference bases. See detailed documentation [here](http://support.illumina.com/help/SequencingAnalysisWorkflow/Content/Vault/Informatics/Sequencing_Analysis/CASAVA/swSEQ_mCA_ExtendedCIGARFormat.htm).
   */
  cigar?: string|undefined;
  _cigar?: fhir.IFhirElement|undefined;
  /**
   * End position of the variant on the reference sequence. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  end?: number|undefined;
  _end?: fhir.IFhirElement|undefined;
  /**
   * An allele is one of a set of coexisting sequence variants of a gene ([SO:0001023](http://www.sequenceontology.org/browser/current_svn/term/SO:0001023)).  Nucleotide(s)/amino acids from start position of sequence to stop position of sequence on the positive (+) strand of the observed  sequence. When the sequence  type is DNA, it should be the sequence on the positive (+) strand. This will lay in the range between variant.start and variant.end.
   */
  observedAllele?: string|undefined;
  _observedAllele?: fhir.IFhirElement|undefined;
  /**
   * An allele is one of a set of coexisting sequence variants of a gene ([SO:0001023](http://www.sequenceontology.org/browser/current_svn/term/SO:0001023)). Nucleotide(s)/amino acids from start position of sequence to stop position of sequence on the positive (+) strand of the reference sequence. When the sequence  type is DNA, it should be the sequence on the positive (+) strand. This will lay in the range between variant.start and variant.end.
   */
  referenceAllele?: string|undefined;
  _referenceAllele?: fhir.IFhirElement|undefined;
  /**
   * Start position of the variant on the  reference sequence. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  start?: number|undefined;
  _start?: fhir.IFhirElement|undefined;
  /**
   * A pointer to an Observation containing variant information.
   */
  variantPointer?: fhir.IReference|undefined;
}
/**
 * Receiver Operator Characteristic (ROC) Curve  to give sensitivity/specificity tradeoff.
 */
export type IMolecularSequenceQualityRoc = fhir.IBackboneElement & {
  /**
   * Calculated fScore if the GQ score threshold was set to "score" field value.
   */
  fMeasure?: number[]|undefined;
  _fMeasure?: fhir.IFhirElement[]|undefined;
  /**
   * The number of false negatives if the GQ score threshold was set to "score" field value.
   */
  numFN?: number[]|undefined;
  _numFN?: fhir.IFhirElement[]|undefined;
  /**
   * The number of false positives if the GQ score threshold was set to "score" field value.
   */
  numFP?: number[]|undefined;
  _numFP?: fhir.IFhirElement[]|undefined;
  /**
   * The number of true positives if the GQ score threshold was set to "score" field value.
   */
  numTP?: number[]|undefined;
  _numTP?: fhir.IFhirElement[]|undefined;
  /**
   * Calculated precision if the GQ score threshold was set to "score" field value.
   */
  precision?: number[]|undefined;
  _precision?: fhir.IFhirElement[]|undefined;
  /**
   * Invidual data point representing the GQ (genotype quality) score threshold.
   */
  score?: number[]|undefined;
  _score?: fhir.IFhirElement[]|undefined;
  /**
   * Calculated sensitivity if the GQ score threshold was set to "score" field value.
   */
  sensitivity?: number[]|undefined;
  _sensitivity?: fhir.IFhirElement[]|undefined;
}
/**
 * An experimental feature attribute that defines the quality of the feature in a quantitative way, such as a phred quality score ([SO:0001686](http://www.sequenceontology.org/browser/current_svn/term/SO:0001686)).
 */
export type IMolecularSequenceQuality = fhir.IBackboneElement & {
  /**
   * End position of the sequence. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  end?: number|undefined;
  _end?: fhir.IFhirElement|undefined;
  /**
   * Harmonic mean of Recall and Precision, computed as: 2 * precision * recall / (precision + recall).
   */
  fScore?: number|undefined;
  _fScore?: fhir.IFhirElement|undefined;
  /**
   * The number of false positives where the non-REF alleles in the Truth and Query Call Sets match (i.e. cases where the truth is 1/1 and the query is 0/1 or similar).
   */
  gtFP?: number|undefined;
  _gtFP?: fhir.IFhirElement|undefined;
  /**
   * Which method is used to get sequence quality.
   */
  method?: fhir.ICodeableConcept|undefined;
  /**
   * QUERY.TP / (QUERY.TP + QUERY.FP).
   */
  precision?: number|undefined;
  _precision?: fhir.IFhirElement|undefined;
  /**
   * False positives, i.e. the number of sites in the Query Call Set for which there is no path through the Truth Call Set that is consistent with this site. Sites with correct variant but incorrect genotype are counted here.
   */
  queryFP?: number|undefined;
  _queryFP?: fhir.IFhirElement|undefined;
  /**
   * True positives, from the perspective of the query data, i.e. the number of sites in the Query Call Set for which there are paths through the Truth Call Set that are consistent with all of the alleles at this site, and for which there is an accurate genotype call for the event.
   */
  queryTP?: number|undefined;
  _queryTP?: fhir.IFhirElement|undefined;
  /**
   * TRUTH.TP / (TRUTH.TP + TRUTH.FN).
   */
  recall?: number|undefined;
  _recall?: fhir.IFhirElement|undefined;
  /**
   * Receiver Operator Characteristic (ROC) Curve  to give sensitivity/specificity tradeoff.
   */
  roc?: fhir.IMolecularSequenceQualityRoc|undefined;
  /**
   * The score of an experimentally derived feature such as a p-value ([SO:0001685](http://www.sequenceontology.org/browser/current_svn/term/SO:0001685)).
   */
  score?: fhir.IQuantity|undefined;
  /**
   * Gold standard sequence used for comparing against.
   */
  standardSequence?: fhir.ICodeableConcept|undefined;
  /**
   * Start position of the sequence. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  start?: number|undefined;
  _start?: fhir.IFhirElement|undefined;
  /**
   * False negatives, i.e. the number of sites in the Truth Call Set for which there is no path through the Query Call Set that is consistent with all of the alleles at this site, or sites for which there is an inaccurate genotype call for the event. Sites with correct variant but incorrect genotype are counted here.
   */
  truthFN?: number|undefined;
  _truthFN?: fhir.IFhirElement|undefined;
  /**
   * True positives, from the perspective of the truth data, i.e. the number of sites in the Truth Call Set for which there are paths through the Query Call Set that are consistent with all of the alleles at this site, and for which there is an accurate genotype call for the event.
   */
  truthTP?: number|undefined;
  _truthTP?: fhir.IFhirElement|undefined;
  /**
   * INDEL / SNP / Undefined variant.
   */
  type: MolecularSequenceQualityTypeEnum|null;
  _type?: fhir.IFhirElement|undefined;
}
/**
 * Configurations of the external repository. The repository shall store target's observedSeq or records related with target's observedSeq.
 */
export type IMolecularSequenceRepository = fhir.IBackboneElement & {
  /**
   * Id of the variant in this external repository. The server will understand how to use this id to call for more info about datasets in external repository.
   */
  datasetId?: string|undefined;
  _datasetId?: fhir.IFhirElement|undefined;
  /**
   * URI of an external repository which contains further details about the genetics data.
   */
  name?: string|undefined;
  _name?: fhir.IFhirElement|undefined;
  /**
   * Id of the read in this external repository.
   */
  readsetId?: string|undefined;
  _readsetId?: fhir.IFhirElement|undefined;
  /**
   * Click and see / RESTful API / Need login to see / RESTful API with authentication / Other ways to see resource.
   */
  type: MolecularSequenceRepositoryTypeEnum|null;
  _type?: fhir.IFhirElement|undefined;
  /**
   * URI of an external repository which contains further details about the genetics data.
   */
  url?: string|undefined;
  _url?: fhir.IFhirElement|undefined;
  /**
   * Id of the variantset in this external repository. The server will understand how to use this id to call for more info about variantsets in external repository.
   */
  variantsetId?: string|undefined;
  _variantsetId?: fhir.IFhirElement|undefined;
}
/**
 * Structural variant outer.
 */
export type IMolecularSequenceStructureVariantOuter = fhir.IBackboneElement & {
  /**
   * Structural variant outer end. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  end?: number|undefined;
  _end?: fhir.IFhirElement|undefined;
  /**
   * Structural variant outer start. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  start?: number|undefined;
  _start?: fhir.IFhirElement|undefined;
}
/**
 * Structural variant inner.
 */
export type IMolecularSequenceStructureVariantInner = fhir.IBackboneElement & {
  /**
   * Structural variant inner end. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  end?: number|undefined;
  _end?: fhir.IFhirElement|undefined;
  /**
   * Structural variant inner start. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  start?: number|undefined;
  _start?: fhir.IFhirElement|undefined;
}
/**
 * Information about chromosome structure variation.
 */
export type IMolecularSequenceStructureVariant = fhir.IBackboneElement & {
  /**
   * Used to indicate if the outer and inner start-end values have the same meaning.
   */
  exact?: boolean|undefined;
  _exact?: fhir.IFhirElement|undefined;
  /**
   * Structural variant inner.
   */
  inner?: fhir.IMolecularSequenceStructureVariantInner|undefined;
  /**
   * Length of the variant chromosome.
   */
  length?: number|undefined;
  _length?: fhir.IFhirElement|undefined;
  /**
   * Structural variant outer.
   */
  outer?: fhir.IMolecularSequenceStructureVariantOuter|undefined;
  /**
   * Information about chromosome structure variation DNA change type.
   */
  variantType?: fhir.ICodeableConcept|undefined;
}
/**
 * Raw data describing a biological sequence.
 */
export type IMolecularSequence = fhir.IDomainResource & {
  /**
   * Resource Type Name
   */
  resourceType: "MolecularSequence";
  /**
   * Whether the sequence is numbered starting at 0 (0-based numbering or coordinates, inclusive start, exclusive end) or starting at 1 (1-based numbering, inclusive start and inclusive end).
   */
  coordinateSystem: number|null;
  _coordinateSystem?: fhir.IFhirElement|undefined;
  /**
   * The method for sequencing, for example, chip information.
   */
  device?: fhir.IReference|undefined;
  /**
   * A unique identifier for this particular sequence instance. This is a FHIR-defined id.
   */
  identifier?: fhir.IIdentifier[]|undefined;
  /**
   * Sequence that was observed. It is the result marked by referenceSeq along with variant records on referenceSeq. This shall start from referenceSeq.windowStart and end by referenceSeq.windowEnd.
   */
  observedSeq?: string|undefined;
  _observedSeq?: fhir.IFhirElement|undefined;
  /**
   * The patient whose sequencing results are described by this resource.
   */
  patient?: fhir.IReference|undefined;
  /**
   * The organization or lab that should be responsible for this result.
   */
  performer?: fhir.IReference|undefined;
  /**
   * Pointer to next atomic sequence which at most contains one variant.
   */
  pointer?: fhir.IReference[]|undefined;
  /**
   * An experimental feature attribute that defines the quality of the feature in a quantitative way, such as a phred quality score ([SO:0001686](http://www.sequenceontology.org/browser/current_svn/term/SO:0001686)).
   */
  quality?: fhir.IMolecularSequenceQuality[]|undefined;
  /**
   * The number of copies of the sequence of interest. (RNASeq).
   */
  quantity?: fhir.IQuantity|undefined;
  /**
   * Coverage (read depth or depth) is the average number of reads representing a given nucleotide in the reconstructed sequence.
   */
  readCoverage?: number|undefined;
  _readCoverage?: fhir.IFhirElement|undefined;
  /**
   * A sequence that is used as a reference to describe variants that are present in a sequence analyzed.
   */
  referenceSeq?: fhir.IMolecularSequenceReferenceSeq|undefined;
  /**
   * Configurations of the external repository. The repository shall store target's observedSeq or records related with target's observedSeq.
   */
  repository?: fhir.IMolecularSequenceRepository[]|undefined;
  /**
   * Specimen used for sequencing.
   */
  specimen?: fhir.IReference|undefined;
  /**
   * Information about chromosome structure variation.
   */
  structureVariant?: fhir.IMolecularSequenceStructureVariant[]|undefined;
  /**
   * Amino Acid Sequence/ DNA Sequence / RNA Sequence.
   */
  type?: MolecularSequenceTypeEnum|undefined;
  _type?: fhir.IFhirElement|undefined;
  /**
   * The definition of variant here originates from Sequence ontology ([variant_of](http://www.sequenceontology.org/browser/current_svn/term/variant_of)). This element can represent amino acid or nucleic sequence change(including insertion,deletion,SNP,etc.)  It can represent some complex mutation or segment variation with the assist of CIGAR string.
   */
  variant?: fhir.IMolecularSequenceVariant[]|undefined;
}
/**
 * A sequence that is used as a reference to describe variants that are present in a sequence analyzed.
 */
export class MolecularSequenceReferenceSeq extends fhir.BackboneElement implements fhir.IMolecularSequenceReferenceSeq {
  /**
   * Structural unit composed of a nucleic acid molecule which controls its own replication through the interaction of specific proteins at one or more origins of replication ([SO:0000340](http://www.sequenceontology.org/browser/current_svn/term/SO:0000340)).
   */
  public chromosome?: fhir.CodeableConcept|undefined;
  /**
   * The Genome Build used for reference, following GRCh build versions e.g. 'GRCh 37'.  Version number must be included if a versioned release of a primary build was used.
   */
  public genomeBuild?: string|undefined;
  public _genomeBuild?: fhir.FhirElement|undefined;
  /**
   * A relative reference to a DNA strand based on gene orientation. The strand that contains the open reading frame of the gene is the "sense" strand, and the opposite complementary strand is the "antisense" strand.
   */
  public orientation?: MolecularSequenceReferenceSeqOrientationEnum|undefined;
  public _orientation?: fhir.FhirElement|undefined;
  /**
   * Reference identifier of reference sequence submitted to NCBI. It must match the type in the MolecularSequence.type field. For example, the prefix, “NG_” identifies reference sequence for genes, “NM_” for messenger RNA transcripts, and “NP_” for amino acid sequences.
   */
  public referenceSeqId?: fhir.CodeableConcept|undefined;
  /**
   * A pointer to another MolecularSequence entity as reference sequence.
   */
  public referenceSeqPointer?: fhir.Reference|undefined;
  /**
   * A string like "ACGT".
   */
  public referenceSeqString?: string|undefined;
  public _referenceSeqString?: fhir.FhirElement|undefined;
  /**
   * An absolute reference to a strand. The Watson strand is the strand whose 5'-end is on the short arm of the chromosome, and the Crick strand as the one whose 5'-end is on the long arm.
   */
  public strand?: MolecularSequenceReferenceSeqStrandEnum|undefined;
  public _strand?: fhir.FhirElement|undefined;
  /**
   * End position of the window on the reference sequence. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  public windowEnd?: number|undefined;
  public _windowEnd?: fhir.FhirElement|undefined;
  /**
   * Start position of the window on the reference sequence. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  public windowStart?: number|undefined;
  public _windowStart?: fhir.FhirElement|undefined;
  /**
   * Default constructor for MolecularSequenceReferenceSeq - initializes any required elements to null if a value is not provided.
   */
  constructor(source:Partial<fhir.IMolecularSequenceReferenceSeq> = {}) {
    super(source);
    if (source["chromosome"]) { this.chromosome = new fhir.CodeableConcept(source.chromosome!); }
    if (source["genomeBuild"]) { this.genomeBuild = source.genomeBuild; }
    if (source["_genomeBuild"]) { this._genomeBuild = new fhir.FhirElement(source._genomeBuild!); }
    if (source["orientation"]) { this.orientation = source.orientation; }
    if (source["_orientation"]) { this._orientation = new fhir.FhirElement(source._orientation!); }
    if (source["referenceSeqId"]) { this.referenceSeqId = new fhir.CodeableConcept(source.referenceSeqId!); }
    if (source["referenceSeqPointer"]) { this.referenceSeqPointer = new fhir.Reference(source.referenceSeqPointer!); }
    if (source["referenceSeqString"]) { this.referenceSeqString = source.referenceSeqString; }
    if (source["_referenceSeqString"]) { this._referenceSeqString = new fhir.FhirElement(source._referenceSeqString!); }
    if (source["strand"]) { this.strand = source.strand; }
    if (source["_strand"]) { this._strand = new fhir.FhirElement(source._strand!); }
    if (source["windowEnd"]) { this.windowEnd = source.windowEnd; }
    if (source["_windowEnd"]) { this._windowEnd = new fhir.FhirElement(source._windowEnd!); }
    if (source["windowStart"]) { this.windowStart = source.windowStart; }
    if (source["_windowStart"]) { this._windowStart = new fhir.FhirElement(source._windowStart!); }
  }
  /**
   * Check if the current MolecularSequenceReferenceSeq contains all required elements.
   */
  override checkRequiredElements():string[] {
    var missingElements:string[] = [];
    var parentMissing:string[] = super.checkRequiredElements();
    missingElements.push(...parentMissing);
    return missingElements;
  }
  /**
   * Factory function to create a MolecularSequenceReferenceSeq from an object that MUST contain all required elements.
   */
  static override fromStrict(source:fhir.IMolecularSequenceReferenceSeq):MolecularSequenceReferenceSeq {
    var dest:MolecularSequenceReferenceSeq = new MolecularSequenceReferenceSeq(source);
    var missingElements:string[] = dest.checkRequiredElements();
    if (missingElements.length !== 0) { throw `MolecularSequenceReferenceSeq is missing elements: ${missingElements.join(", ")}` }
    return dest;
  }
}
/**
 * The definition of variant here originates from Sequence ontology ([variant_of](http://www.sequenceontology.org/browser/current_svn/term/variant_of)). This element can represent amino acid or nucleic sequence change(including insertion,deletion,SNP,etc.)  It can represent some complex mutation or segment variation with the assist of CIGAR string.
 */
export class MolecularSequenceVariant extends fhir.BackboneElement implements fhir.IMolecularSequenceVariant {
  /**
   * Extended CIGAR string for aligning the sequence with reference bases. See detailed documentation [here](http://support.illumina.com/help/SequencingAnalysisWorkflow/Content/Vault/Informatics/Sequencing_Analysis/CASAVA/swSEQ_mCA_ExtendedCIGARFormat.htm).
   */
  public cigar?: string|undefined;
  public _cigar?: fhir.FhirElement|undefined;
  /**
   * End position of the variant on the reference sequence. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  public end?: number|undefined;
  public _end?: fhir.FhirElement|undefined;
  /**
   * An allele is one of a set of coexisting sequence variants of a gene ([SO:0001023](http://www.sequenceontology.org/browser/current_svn/term/SO:0001023)).  Nucleotide(s)/amino acids from start position of sequence to stop position of sequence on the positive (+) strand of the observed  sequence. When the sequence  type is DNA, it should be the sequence on the positive (+) strand. This will lay in the range between variant.start and variant.end.
   */
  public observedAllele?: string|undefined;
  public _observedAllele?: fhir.FhirElement|undefined;
  /**
   * An allele is one of a set of coexisting sequence variants of a gene ([SO:0001023](http://www.sequenceontology.org/browser/current_svn/term/SO:0001023)). Nucleotide(s)/amino acids from start position of sequence to stop position of sequence on the positive (+) strand of the reference sequence. When the sequence  type is DNA, it should be the sequence on the positive (+) strand. This will lay in the range between variant.start and variant.end.
   */
  public referenceAllele?: string|undefined;
  public _referenceAllele?: fhir.FhirElement|undefined;
  /**
   * Start position of the variant on the  reference sequence. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  public start?: number|undefined;
  public _start?: fhir.FhirElement|undefined;
  /**
   * A pointer to an Observation containing variant information.
   */
  public variantPointer?: fhir.Reference|undefined;
  /**
   * Default constructor for MolecularSequenceVariant - initializes any required elements to null if a value is not provided.
   */
  constructor(source:Partial<fhir.IMolecularSequenceVariant> = {}) {
    super(source);
    if (source["cigar"]) { this.cigar = source.cigar; }
    if (source["_cigar"]) { this._cigar = new fhir.FhirElement(source._cigar!); }
    if (source["end"]) { this.end = source.end; }
    if (source["_end"]) { this._end = new fhir.FhirElement(source._end!); }
    if (source["observedAllele"]) { this.observedAllele = source.observedAllele; }
    if (source["_observedAllele"]) { this._observedAllele = new fhir.FhirElement(source._observedAllele!); }
    if (source["referenceAllele"]) { this.referenceAllele = source.referenceAllele; }
    if (source["_referenceAllele"]) { this._referenceAllele = new fhir.FhirElement(source._referenceAllele!); }
    if (source["start"]) { this.start = source.start; }
    if (source["_start"]) { this._start = new fhir.FhirElement(source._start!); }
    if (source["variantPointer"]) { this.variantPointer = new fhir.Reference(source.variantPointer!); }
  }
  /**
   * Check if the current MolecularSequenceVariant contains all required elements.
   */
  override checkRequiredElements():string[] {
    var missingElements:string[] = [];
    var parentMissing:string[] = super.checkRequiredElements();
    missingElements.push(...parentMissing);
    return missingElements;
  }
  /**
   * Factory function to create a MolecularSequenceVariant from an object that MUST contain all required elements.
   */
  static override fromStrict(source:fhir.IMolecularSequenceVariant):MolecularSequenceVariant {
    var dest:MolecularSequenceVariant = new MolecularSequenceVariant(source);
    var missingElements:string[] = dest.checkRequiredElements();
    if (missingElements.length !== 0) { throw `MolecularSequenceVariant is missing elements: ${missingElements.join(", ")}` }
    return dest;
  }
}
/**
 * Receiver Operator Characteristic (ROC) Curve  to give sensitivity/specificity tradeoff.
 */
export class MolecularSequenceQualityRoc extends fhir.BackboneElement implements fhir.IMolecularSequenceQualityRoc {
  /**
   * Calculated fScore if the GQ score threshold was set to "score" field value.
   */
  public fMeasure?: number[]|undefined;
  public _fMeasure?: fhir.FhirElement[]|undefined;
  /**
   * The number of false negatives if the GQ score threshold was set to "score" field value.
   */
  public numFN?: number[]|undefined;
  public _numFN?: fhir.FhirElement[]|undefined;
  /**
   * The number of false positives if the GQ score threshold was set to "score" field value.
   */
  public numFP?: number[]|undefined;
  public _numFP?: fhir.FhirElement[]|undefined;
  /**
   * The number of true positives if the GQ score threshold was set to "score" field value.
   */
  public numTP?: number[]|undefined;
  public _numTP?: fhir.FhirElement[]|undefined;
  /**
   * Calculated precision if the GQ score threshold was set to "score" field value.
   */
  public precision?: number[]|undefined;
  public _precision?: fhir.FhirElement[]|undefined;
  /**
   * Invidual data point representing the GQ (genotype quality) score threshold.
   */
  public score?: number[]|undefined;
  public _score?: fhir.FhirElement[]|undefined;
  /**
   * Calculated sensitivity if the GQ score threshold was set to "score" field value.
   */
  public sensitivity?: number[]|undefined;
  public _sensitivity?: fhir.FhirElement[]|undefined;
  /**
   * Default constructor for MolecularSequenceQualityRoc - initializes any required elements to null if a value is not provided.
   */
  constructor(source:Partial<fhir.IMolecularSequenceQualityRoc> = {}) {
    super(source);
    if (source["fMeasure"]) { this.fMeasure = source.fMeasure.map((x) => (x)); }
    if (source["_fMeasure"]) { this._fMeasure = source._fMeasure.map((x:Partial<fhir.IFhirElement>) => new fhir.FhirElement(x)); }
    if (source["numFN"]) { this.numFN = source.numFN.map((x) => (x)); }
    if (source["_numFN"]) { this._numFN = source._numFN.map((x:Partial<fhir.IFhirElement>) => new fhir.FhirElement(x)); }
    if (source["numFP"]) { this.numFP = source.numFP.map((x) => (x)); }
    if (source["_numFP"]) { this._numFP = source._numFP.map((x:Partial<fhir.IFhirElement>) => new fhir.FhirElement(x)); }
    if (source["numTP"]) { this.numTP = source.numTP.map((x) => (x)); }
    if (source["_numTP"]) { this._numTP = source._numTP.map((x:Partial<fhir.IFhirElement>) => new fhir.FhirElement(x)); }
    if (source["precision"]) { this.precision = source.precision.map((x) => (x)); }
    if (source["_precision"]) { this._precision = source._precision.map((x:Partial<fhir.IFhirElement>) => new fhir.FhirElement(x)); }
    if (source["score"]) { this.score = source.score.map((x) => (x)); }
    if (source["_score"]) { this._score = source._score.map((x:Partial<fhir.IFhirElement>) => new fhir.FhirElement(x)); }
    if (source["sensitivity"]) { this.sensitivity = source.sensitivity.map((x) => (x)); }
    if (source["_sensitivity"]) { this._sensitivity = source._sensitivity.map((x:Partial<fhir.IFhirElement>) => new fhir.FhirElement(x)); }
  }
  /**
   * Check if the current MolecularSequenceQualityRoc contains all required elements.
   */
  override checkRequiredElements():string[] {
    var missingElements:string[] = [];
    var parentMissing:string[] = super.checkRequiredElements();
    missingElements.push(...parentMissing);
    return missingElements;
  }
  /**
   * Factory function to create a MolecularSequenceQualityRoc from an object that MUST contain all required elements.
   */
  static override fromStrict(source:fhir.IMolecularSequenceQualityRoc):MolecularSequenceQualityRoc {
    var dest:MolecularSequenceQualityRoc = new MolecularSequenceQualityRoc(source);
    var missingElements:string[] = dest.checkRequiredElements();
    if (missingElements.length !== 0) { throw `MolecularSequenceQualityRoc is missing elements: ${missingElements.join(", ")}` }
    return dest;
  }
}
/**
 * An experimental feature attribute that defines the quality of the feature in a quantitative way, such as a phred quality score ([SO:0001686](http://www.sequenceontology.org/browser/current_svn/term/SO:0001686)).
 */
export class MolecularSequenceQuality extends fhir.BackboneElement implements fhir.IMolecularSequenceQuality {
  /**
   * End position of the sequence. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  public end?: number|undefined;
  public _end?: fhir.FhirElement|undefined;
  /**
   * Harmonic mean of Recall and Precision, computed as: 2 * precision * recall / (precision + recall).
   */
  public fScore?: number|undefined;
  public _fScore?: fhir.FhirElement|undefined;
  /**
   * The number of false positives where the non-REF alleles in the Truth and Query Call Sets match (i.e. cases where the truth is 1/1 and the query is 0/1 or similar).
   */
  public gtFP?: number|undefined;
  public _gtFP?: fhir.FhirElement|undefined;
  /**
   * Which method is used to get sequence quality.
   */
  public method?: fhir.CodeableConcept|undefined;
  /**
   * QUERY.TP / (QUERY.TP + QUERY.FP).
   */
  public precision?: number|undefined;
  public _precision?: fhir.FhirElement|undefined;
  /**
   * False positives, i.e. the number of sites in the Query Call Set for which there is no path through the Truth Call Set that is consistent with this site. Sites with correct variant but incorrect genotype are counted here.
   */
  public queryFP?: number|undefined;
  public _queryFP?: fhir.FhirElement|undefined;
  /**
   * True positives, from the perspective of the query data, i.e. the number of sites in the Query Call Set for which there are paths through the Truth Call Set that are consistent with all of the alleles at this site, and for which there is an accurate genotype call for the event.
   */
  public queryTP?: number|undefined;
  public _queryTP?: fhir.FhirElement|undefined;
  /**
   * TRUTH.TP / (TRUTH.TP + TRUTH.FN).
   */
  public recall?: number|undefined;
  public _recall?: fhir.FhirElement|undefined;
  /**
   * Receiver Operator Characteristic (ROC) Curve  to give sensitivity/specificity tradeoff.
   */
  public roc?: fhir.MolecularSequenceQualityRoc|undefined;
  /**
   * The score of an experimentally derived feature such as a p-value ([SO:0001685](http://www.sequenceontology.org/browser/current_svn/term/SO:0001685)).
   */
  public score?: fhir.Quantity|undefined;
  /**
   * Gold standard sequence used for comparing against.
   */
  public standardSequence?: fhir.CodeableConcept|undefined;
  /**
   * Start position of the sequence. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  public start?: number|undefined;
  public _start?: fhir.FhirElement|undefined;
  /**
   * False negatives, i.e. the number of sites in the Truth Call Set for which there is no path through the Query Call Set that is consistent with all of the alleles at this site, or sites for which there is an inaccurate genotype call for the event. Sites with correct variant but incorrect genotype are counted here.
   */
  public truthFN?: number|undefined;
  public _truthFN?: fhir.FhirElement|undefined;
  /**
   * True positives, from the perspective of the truth data, i.e. the number of sites in the Truth Call Set for which there are paths through the Query Call Set that are consistent with all of the alleles at this site, and for which there is an accurate genotype call for the event.
   */
  public truthTP?: number|undefined;
  public _truthTP?: fhir.FhirElement|undefined;
  /**
   * INDEL / SNP / Undefined variant.
   */
  public type: MolecularSequenceQualityTypeEnum|null;
  public _type?: fhir.FhirElement|undefined;
  /**
   * Default constructor for MolecularSequenceQuality - initializes any required elements to null if a value is not provided.
   */
  constructor(source:Partial<fhir.IMolecularSequenceQuality> = {}) {
    super(source);
    if (source["end"]) { this.end = source.end; }
    if (source["_end"]) { this._end = new fhir.FhirElement(source._end!); }
    if (source["fScore"]) { this.fScore = source.fScore; }
    if (source["_fScore"]) { this._fScore = new fhir.FhirElement(source._fScore!); }
    if (source["gtFP"]) { this.gtFP = source.gtFP; }
    if (source["_gtFP"]) { this._gtFP = new fhir.FhirElement(source._gtFP!); }
    if (source["method"]) { this.method = new fhir.CodeableConcept(source.method!); }
    if (source["precision"]) { this.precision = source.precision; }
    if (source["_precision"]) { this._precision = new fhir.FhirElement(source._precision!); }
    if (source["queryFP"]) { this.queryFP = source.queryFP; }
    if (source["_queryFP"]) { this._queryFP = new fhir.FhirElement(source._queryFP!); }
    if (source["queryTP"]) { this.queryTP = source.queryTP; }
    if (source["_queryTP"]) { this._queryTP = new fhir.FhirElement(source._queryTP!); }
    if (source["recall"]) { this.recall = source.recall; }
    if (source["_recall"]) { this._recall = new fhir.FhirElement(source._recall!); }
    if (source["roc"]) { this.roc = new fhir.MolecularSequenceQualityRoc(source.roc!); }
    if (source["score"]) { this.score = new fhir.Quantity(source.score!); }
    if (source["standardSequence"]) { this.standardSequence = new fhir.CodeableConcept(source.standardSequence!); }
    if (source["start"]) { this.start = source.start; }
    if (source["_start"]) { this._start = new fhir.FhirElement(source._start!); }
    if (source["truthFN"]) { this.truthFN = source.truthFN; }
    if (source["_truthFN"]) { this._truthFN = new fhir.FhirElement(source._truthFN!); }
    if (source["truthTP"]) { this.truthTP = source.truthTP; }
    if (source["_truthTP"]) { this._truthTP = new fhir.FhirElement(source._truthTP!); }
    this.type = null;
    if (source["type"]) { this.type = source.type; }
    if (this.type === undefined) { this.type = null }
    if (source["_type"]) { this._type = new fhir.FhirElement(source._type!); }
  }
  /**
   * Check if the current MolecularSequenceQuality contains all required elements.
   */
  override checkRequiredElements():string[] {
    var missingElements:string[] = [];
    if (this["type"] === undefined) { missingElements.push("type"); }
    var parentMissing:string[] = super.checkRequiredElements();
    missingElements.push(...parentMissing);
    return missingElements;
  }
  /**
   * Factory function to create a MolecularSequenceQuality from an object that MUST contain all required elements.
   */
  static override fromStrict(source:fhir.IMolecularSequenceQuality):MolecularSequenceQuality {
    var dest:MolecularSequenceQuality = new MolecularSequenceQuality(source);
    var missingElements:string[] = dest.checkRequiredElements();
    if (missingElements.length !== 0) { throw `MolecularSequenceQuality is missing elements: ${missingElements.join(", ")}` }
    return dest;
  }
}
/**
 * Configurations of the external repository. The repository shall store target's observedSeq or records related with target's observedSeq.
 */
export class MolecularSequenceRepository extends fhir.BackboneElement implements fhir.IMolecularSequenceRepository {
  /**
   * Id of the variant in this external repository. The server will understand how to use this id to call for more info about datasets in external repository.
   */
  public datasetId?: string|undefined;
  public _datasetId?: fhir.FhirElement|undefined;
  /**
   * URI of an external repository which contains further details about the genetics data.
   */
  public name?: string|undefined;
  public _name?: fhir.FhirElement|undefined;
  /**
   * Id of the read in this external repository.
   */
  public readsetId?: string|undefined;
  public _readsetId?: fhir.FhirElement|undefined;
  /**
   * Click and see / RESTful API / Need login to see / RESTful API with authentication / Other ways to see resource.
   */
  public type: MolecularSequenceRepositoryTypeEnum|null;
  public _type?: fhir.FhirElement|undefined;
  /**
   * URI of an external repository which contains further details about the genetics data.
   */
  public url?: string|undefined;
  public _url?: fhir.FhirElement|undefined;
  /**
   * Id of the variantset in this external repository. The server will understand how to use this id to call for more info about variantsets in external repository.
   */
  public variantsetId?: string|undefined;
  public _variantsetId?: fhir.FhirElement|undefined;
  /**
   * Default constructor for MolecularSequenceRepository - initializes any required elements to null if a value is not provided.
   */
  constructor(source:Partial<fhir.IMolecularSequenceRepository> = {}) {
    super(source);
    if (source["datasetId"]) { this.datasetId = source.datasetId; }
    if (source["_datasetId"]) { this._datasetId = new fhir.FhirElement(source._datasetId!); }
    if (source["name"]) { this.name = source.name; }
    if (source["_name"]) { this._name = new fhir.FhirElement(source._name!); }
    if (source["readsetId"]) { this.readsetId = source.readsetId; }
    if (source["_readsetId"]) { this._readsetId = new fhir.FhirElement(source._readsetId!); }
    this.type = null;
    if (source["type"]) { this.type = source.type; }
    if (this.type === undefined) { this.type = null }
    if (source["_type"]) { this._type = new fhir.FhirElement(source._type!); }
    if (source["url"]) { this.url = source.url; }
    if (source["_url"]) { this._url = new fhir.FhirElement(source._url!); }
    if (source["variantsetId"]) { this.variantsetId = source.variantsetId; }
    if (source["_variantsetId"]) { this._variantsetId = new fhir.FhirElement(source._variantsetId!); }
  }
  /**
   * Check if the current MolecularSequenceRepository contains all required elements.
   */
  override checkRequiredElements():string[] {
    var missingElements:string[] = [];
    if (this["type"] === undefined) { missingElements.push("type"); }
    var parentMissing:string[] = super.checkRequiredElements();
    missingElements.push(...parentMissing);
    return missingElements;
  }
  /**
   * Factory function to create a MolecularSequenceRepository from an object that MUST contain all required elements.
   */
  static override fromStrict(source:fhir.IMolecularSequenceRepository):MolecularSequenceRepository {
    var dest:MolecularSequenceRepository = new MolecularSequenceRepository(source);
    var missingElements:string[] = dest.checkRequiredElements();
    if (missingElements.length !== 0) { throw `MolecularSequenceRepository is missing elements: ${missingElements.join(", ")}` }
    return dest;
  }
}
/**
 * Structural variant outer.
 */
export class MolecularSequenceStructureVariantOuter extends fhir.BackboneElement implements fhir.IMolecularSequenceStructureVariantOuter {
  /**
   * Structural variant outer end. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  public end?: number|undefined;
  public _end?: fhir.FhirElement|undefined;
  /**
   * Structural variant outer start. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  public start?: number|undefined;
  public _start?: fhir.FhirElement|undefined;
  /**
   * Default constructor for MolecularSequenceStructureVariantOuter - initializes any required elements to null if a value is not provided.
   */
  constructor(source:Partial<fhir.IMolecularSequenceStructureVariantOuter> = {}) {
    super(source);
    if (source["end"]) { this.end = source.end; }
    if (source["_end"]) { this._end = new fhir.FhirElement(source._end!); }
    if (source["start"]) { this.start = source.start; }
    if (source["_start"]) { this._start = new fhir.FhirElement(source._start!); }
  }
  /**
   * Check if the current MolecularSequenceStructureVariantOuter contains all required elements.
   */
  override checkRequiredElements():string[] {
    var missingElements:string[] = [];
    var parentMissing:string[] = super.checkRequiredElements();
    missingElements.push(...parentMissing);
    return missingElements;
  }
  /**
   * Factory function to create a MolecularSequenceStructureVariantOuter from an object that MUST contain all required elements.
   */
  static override fromStrict(source:fhir.IMolecularSequenceStructureVariantOuter):MolecularSequenceStructureVariantOuter {
    var dest:MolecularSequenceStructureVariantOuter = new MolecularSequenceStructureVariantOuter(source);
    var missingElements:string[] = dest.checkRequiredElements();
    if (missingElements.length !== 0) { throw `MolecularSequenceStructureVariantOuter is missing elements: ${missingElements.join(", ")}` }
    return dest;
  }
}
/**
 * Structural variant inner.
 */
export class MolecularSequenceStructureVariantInner extends fhir.BackboneElement implements fhir.IMolecularSequenceStructureVariantInner {
  /**
   * Structural variant inner end. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  public end?: number|undefined;
  public _end?: fhir.FhirElement|undefined;
  /**
   * Structural variant inner start. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  public start?: number|undefined;
  public _start?: fhir.FhirElement|undefined;
  /**
   * Default constructor for MolecularSequenceStructureVariantInner - initializes any required elements to null if a value is not provided.
   */
  constructor(source:Partial<fhir.IMolecularSequenceStructureVariantInner> = {}) {
    super(source);
    if (source["end"]) { this.end = source.end; }
    if (source["_end"]) { this._end = new fhir.FhirElement(source._end!); }
    if (source["start"]) { this.start = source.start; }
    if (source["_start"]) { this._start = new fhir.FhirElement(source._start!); }
  }
  /**
   * Check if the current MolecularSequenceStructureVariantInner contains all required elements.
   */
  override checkRequiredElements():string[] {
    var missingElements:string[] = [];
    var parentMissing:string[] = super.checkRequiredElements();
    missingElements.push(...parentMissing);
    return missingElements;
  }
  /**
   * Factory function to create a MolecularSequenceStructureVariantInner from an object that MUST contain all required elements.
   */
  static override fromStrict(source:fhir.IMolecularSequenceStructureVariantInner):MolecularSequenceStructureVariantInner {
    var dest:MolecularSequenceStructureVariantInner = new MolecularSequenceStructureVariantInner(source);
    var missingElements:string[] = dest.checkRequiredElements();
    if (missingElements.length !== 0) { throw `MolecularSequenceStructureVariantInner is missing elements: ${missingElements.join(", ")}` }
    return dest;
  }
}
/**
 * Information about chromosome structure variation.
 */
export class MolecularSequenceStructureVariant extends fhir.BackboneElement implements fhir.IMolecularSequenceStructureVariant {
  /**
   * Used to indicate if the outer and inner start-end values have the same meaning.
   */
  public exact?: boolean|undefined;
  public _exact?: fhir.FhirElement|undefined;
  /**
   * Structural variant inner.
   */
  public inner?: fhir.MolecularSequenceStructureVariantInner|undefined;
  /**
   * Length of the variant chromosome.
   */
  public length?: number|undefined;
  public _length?: fhir.FhirElement|undefined;
  /**
   * Structural variant outer.
   */
  public outer?: fhir.MolecularSequenceStructureVariantOuter|undefined;
  /**
   * Information about chromosome structure variation DNA change type.
   */
  public variantType?: fhir.CodeableConcept|undefined;
  /**
   * Default constructor for MolecularSequenceStructureVariant - initializes any required elements to null if a value is not provided.
   */
  constructor(source:Partial<fhir.IMolecularSequenceStructureVariant> = {}) {
    super(source);
    if (source["exact"]) { this.exact = source.exact; }
    if (source["_exact"]) { this._exact = new fhir.FhirElement(source._exact!); }
    if (source["inner"]) { this.inner = new fhir.MolecularSequenceStructureVariantInner(source.inner!); }
    if (source["length"]) { this.length = source.length; }
    if (source["_length"]) { this._length = new fhir.FhirElement(source._length!); }
    if (source["outer"]) { this.outer = new fhir.MolecularSequenceStructureVariantOuter(source.outer!); }
    if (source["variantType"]) { this.variantType = new fhir.CodeableConcept(source.variantType!); }
  }
  /**
   * Check if the current MolecularSequenceStructureVariant contains all required elements.
   */
  override checkRequiredElements():string[] {
    var missingElements:string[] = [];
    var parentMissing:string[] = super.checkRequiredElements();
    missingElements.push(...parentMissing);
    return missingElements;
  }
  /**
   * Factory function to create a MolecularSequenceStructureVariant from an object that MUST contain all required elements.
   */
  static override fromStrict(source:fhir.IMolecularSequenceStructureVariant):MolecularSequenceStructureVariant {
    var dest:MolecularSequenceStructureVariant = new MolecularSequenceStructureVariant(source);
    var missingElements:string[] = dest.checkRequiredElements();
    if (missingElements.length !== 0) { throw `MolecularSequenceStructureVariant is missing elements: ${missingElements.join(", ")}` }
    return dest;
  }
}
/**
 * Raw data describing a biological sequence.
 */
export class MolecularSequence extends fhir.DomainResource implements fhir.IMolecularSequence {
  /**
   * Resource Type Name
   */
  public override resourceType: "MolecularSequence";
  /**
   * Whether the sequence is numbered starting at 0 (0-based numbering or coordinates, inclusive start, exclusive end) or starting at 1 (1-based numbering, inclusive start and inclusive end).
   */
  public coordinateSystem: number|null;
  public _coordinateSystem?: fhir.FhirElement|undefined;
  /**
   * The method for sequencing, for example, chip information.
   */
  public device?: fhir.Reference|undefined;
  /**
   * A unique identifier for this particular sequence instance. This is a FHIR-defined id.
   */
  public identifier?: fhir.Identifier[]|undefined;
  /**
   * Sequence that was observed. It is the result marked by referenceSeq along with variant records on referenceSeq. This shall start from referenceSeq.windowStart and end by referenceSeq.windowEnd.
   */
  public observedSeq?: string|undefined;
  public _observedSeq?: fhir.FhirElement|undefined;
  /**
   * The patient whose sequencing results are described by this resource.
   */
  public patient?: fhir.Reference|undefined;
  /**
   * The organization or lab that should be responsible for this result.
   */
  public performer?: fhir.Reference|undefined;
  /**
   * Pointer to next atomic sequence which at most contains one variant.
   */
  public pointer?: fhir.Reference[]|undefined;
  /**
   * An experimental feature attribute that defines the quality of the feature in a quantitative way, such as a phred quality score ([SO:0001686](http://www.sequenceontology.org/browser/current_svn/term/SO:0001686)).
   */
  public quality?: fhir.MolecularSequenceQuality[]|undefined;
  /**
   * The number of copies of the sequence of interest. (RNASeq).
   */
  public quantity?: fhir.Quantity|undefined;
  /**
   * Coverage (read depth or depth) is the average number of reads representing a given nucleotide in the reconstructed sequence.
   */
  public readCoverage?: number|undefined;
  public _readCoverage?: fhir.FhirElement|undefined;
  /**
   * A sequence that is used as a reference to describe variants that are present in a sequence analyzed.
   */
  public referenceSeq?: fhir.MolecularSequenceReferenceSeq|undefined;
  /**
   * Configurations of the external repository. The repository shall store target's observedSeq or records related with target's observedSeq.
   */
  public repository?: fhir.MolecularSequenceRepository[]|undefined;
  /**
   * Specimen used for sequencing.
   */
  public specimen?: fhir.Reference|undefined;
  /**
   * Information about chromosome structure variation.
   */
  public structureVariant?: fhir.MolecularSequenceStructureVariant[]|undefined;
  /**
   * Amino Acid Sequence/ DNA Sequence / RNA Sequence.
   */
  public type?: MolecularSequenceTypeEnum|undefined;
  public _type?: fhir.FhirElement|undefined;
  /**
   * The definition of variant here originates from Sequence ontology ([variant_of](http://www.sequenceontology.org/browser/current_svn/term/variant_of)). This element can represent amino acid or nucleic sequence change(including insertion,deletion,SNP,etc.)  It can represent some complex mutation or segment variation with the assist of CIGAR string.
   */
  public variant?: fhir.MolecularSequenceVariant[]|undefined;
  /**
   * Default constructor for MolecularSequence - initializes any required elements to null if a value is not provided.
   */
  constructor(source:Partial<fhir.IMolecularSequence> = {}) {
    super(source);
    this.resourceType = 'MolecularSequence';
    this.coordinateSystem = null;
    if (source["coordinateSystem"]) { this.coordinateSystem = source.coordinateSystem; }
    if (this.coordinateSystem === undefined) { this.coordinateSystem = null }
    if (source["_coordinateSystem"]) { this._coordinateSystem = new fhir.FhirElement(source._coordinateSystem!); }
    if (source["device"]) { this.device = new fhir.Reference(source.device!); }
    if (source["identifier"]) { this.identifier = source.identifier.map((x:Partial<fhir.IIdentifier>) => new fhir.Identifier(x)); }
    if (source["observedSeq"]) { this.observedSeq = source.observedSeq; }
    if (source["_observedSeq"]) { this._observedSeq = new fhir.FhirElement(source._observedSeq!); }
    if (source["patient"]) { this.patient = new fhir.Reference(source.patient!); }
    if (source["performer"]) { this.performer = new fhir.Reference(source.performer!); }
    if (source["pointer"]) { this.pointer = source.pointer.map((x:Partial<fhir.IReference>) => new fhir.Reference(x)); }
    if (source["quality"]) { this.quality = source.quality.map((x:Partial<fhir.IMolecularSequenceQuality>) => new fhir.MolecularSequenceQuality(x)); }
    if (source["quantity"]) { this.quantity = new fhir.Quantity(source.quantity!); }
    if (source["readCoverage"]) { this.readCoverage = source.readCoverage; }
    if (source["_readCoverage"]) { this._readCoverage = new fhir.FhirElement(source._readCoverage!); }
    if (source["referenceSeq"]) { this.referenceSeq = new fhir.MolecularSequenceReferenceSeq(source.referenceSeq!); }
    if (source["repository"]) { this.repository = source.repository.map((x:Partial<fhir.IMolecularSequenceRepository>) => new fhir.MolecularSequenceRepository(x)); }
    if (source["specimen"]) { this.specimen = new fhir.Reference(source.specimen!); }
    if (source["structureVariant"]) { this.structureVariant = source.structureVariant.map((x:Partial<fhir.IMolecularSequenceStructureVariant>) => new fhir.MolecularSequenceStructureVariant(x)); }
    if (source["type"]) { this.type = source.type; }
    if (source["_type"]) { this._type = new fhir.FhirElement(source._type!); }
    if (source["variant"]) { this.variant = source.variant.map((x:Partial<fhir.IMolecularSequenceVariant>) => new fhir.MolecularSequenceVariant(x)); }
  }
  /**
   * Check if the current MolecularSequence contains all required elements.
   */
  override checkRequiredElements():string[] {
    var missingElements:string[] = [];
    if (this["coordinateSystem"] === undefined) { missingElements.push("coordinateSystem"); }
    var parentMissing:string[] = super.checkRequiredElements();
    missingElements.push(...parentMissing);
    return missingElements;
  }
  /**
   * Factory function to create a MolecularSequence from an object that MUST contain all required elements.
   */
  static override fromStrict(source:fhir.IMolecularSequence):MolecularSequence {
    var dest:MolecularSequence = new MolecularSequence(source);
    var missingElements:string[] = dest.checkRequiredElements();
    if (missingElements.length !== 0) { throw `MolecularSequence is missing elements: ${missingElements.join(", ")}` }
    return dest;
  }
}
/**
 * Code Values for the MolecularSequence.referenceSeq.orientation field
 */
export enum MolecularSequenceReferenceSeqOrientationEnum {
  SENSE = "sense",
  ANTISENSE = "antisense",
}
/**
 * Code Values for the MolecularSequence.referenceSeq.strand field
 */
export enum MolecularSequenceReferenceSeqStrandEnum {
  WATSON = "watson",
  CRICK = "crick",
}
/**
 * Code Values for the MolecularSequence.quality.type field
 */
export enum MolecularSequenceQualityTypeEnum {
  INDEL = "indel",
  SNP = "snp",
  UNKNOWN = "unknown",
}
/**
 * Code Values for the MolecularSequence.repository.type field
 */
export enum MolecularSequenceRepositoryTypeEnum {
  DIRECTLINK = "directlink",
  OPENAPI = "openapi",
  LOGIN = "login",
  OAUTH = "oauth",
  OTHER = "other",
}
/**
 * Code Values for the MolecularSequence.type field
 */
export enum MolecularSequenceTypeEnum {
  AA = "aa",
  DNA = "dna",
  RNA = "rna",
}
