// <auto-generated/>
// Contents of: hl7.fhir.r4.core version: 4.0.1
// Minimum TypeScript Version: 3.7
// FHIR Resource: MolecularSequence

import * as fhir from '../fhir.js'

import { ChromosomeHumanValueSet, ChromosomeHumanValueSetType, ChromosomeHumanValueSetEnum } from '../fhirValueSets/ChromosomeHumanValueSet.js'
import { OrientationTypeValueSet, OrientationTypeValueSetType, OrientationTypeValueSetEnum } from '../fhirValueSets/OrientationTypeValueSet.js'
import { StrandTypeValueSet, StrandTypeValueSetType, StrandTypeValueSetEnum } from '../fhirValueSets/StrandTypeValueSet.js'
import { QualityTypeValueSet, QualityTypeValueSetType, QualityTypeValueSetEnum } from '../fhirValueSets/QualityTypeValueSet.js'
import { RepositoryTypeValueSet, RepositoryTypeValueSetType, RepositoryTypeValueSetEnum } from '../fhirValueSets/RepositoryTypeValueSet.js'
import { SequenceTypeValueSet, SequenceTypeValueSetType, SequenceTypeValueSetEnum } from '../fhirValueSets/SequenceTypeValueSet.js'

/**
 * A sequence that is used as a reference to describe variants that are present in a sequence analyzed.
 */
export type IMolecularSequenceReferenceSeq = fhir.IBackboneElement & { 
  /**
   * Structural unit composed of a nucleic acid molecule which controls its own replication through the interaction of specific proteins at one or more origins of replication ([SO:0000340](http://www.sequenceontology.org/browser/current_svn/term/SO:0000340)).
   */
  chromosome?: fhir.ICodeableConcept|undefined;
  /**
   * The Genome Build used for reference, following GRCh build versions e.g. 'GRCh 37'.  Version number must be included if a versioned release of a primary build was used.
   */
  genomeBuild?: string|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.referenceSeq.genomeBuild
   */
  _genomeBuild?: fhir.IFhirElement|undefined;
  /**
   * A relative reference to a DNA strand based on gene orientation. The strand that contains the open reading frame of the gene is the "sense" strand, and the opposite complementary strand is the "antisense" strand.
   */
  orientation?: OrientationTypeValueSetEnum|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.referenceSeq.orientation
   */
  _orientation?: fhir.IFhirElement|undefined;
  /**
   * Reference identifier of reference sequence submitted to NCBI. It must match the type in the MolecularSequence.type field. For example, the prefix, “NG_” identifies reference sequence for genes, “NM_” for messenger RNA transcripts, and “NP_” for amino acid sequences.
   */
  referenceSeqId?: fhir.ICodeableConcept|undefined;
  /**
   * A pointer to another MolecularSequence entity as reference sequence.
   */
  referenceSeqPointer?: fhir.IReference|undefined;
  /**
   * A string like "ACGT".
   */
  referenceSeqString?: string|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.referenceSeq.referenceSeqString
   */
  _referenceSeqString?: fhir.IFhirElement|undefined;
  /**
   * An absolute reference to a strand. The Watson strand is the strand whose 5'-end is on the short arm of the chromosome, and the Crick strand as the one whose 5'-end is on the long arm.
   */
  strand?: StrandTypeValueSetEnum|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.referenceSeq.strand
   */
  _strand?: fhir.IFhirElement|undefined;
  /**
   * Start position of the window on the reference sequence. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  windowStart?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.referenceSeq.windowStart
   */
  _windowStart?: fhir.IFhirElement|undefined;
  /**
   * End position of the window on the reference sequence. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  windowEnd?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.referenceSeq.windowEnd
   */
  _windowEnd?: fhir.IFhirElement|undefined;
}

/**
 * The definition of variant here originates from Sequence ontology ([variant_of](http://www.sequenceontology.org/browser/current_svn/term/variant_of)). This element can represent amino acid or nucleic sequence change(including insertion,deletion,SNP,etc.)  It can represent some complex mutation or segment variation with the assist of CIGAR string.
 */
export type IMolecularSequenceVariant = fhir.IBackboneElement & { 
  /**
   * Start position of the variant on the  reference sequence. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  start?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.variant.start
   */
  _start?: fhir.IFhirElement|undefined;
  /**
   * End position of the variant on the reference sequence. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  end?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.variant.end
   */
  _end?: fhir.IFhirElement|undefined;
  /**
   * An allele is one of a set of coexisting sequence variants of a gene ([SO:0001023](http://www.sequenceontology.org/browser/current_svn/term/SO:0001023)).  Nucleotide(s)/amino acids from start position of sequence to stop position of sequence on the positive (+) strand of the observed  sequence. When the sequence  type is DNA, it should be the sequence on the positive (+) strand. This will lay in the range between variant.start and variant.end.
   */
  observedAllele?: string|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.variant.observedAllele
   */
  _observedAllele?: fhir.IFhirElement|undefined;
  /**
   * An allele is one of a set of coexisting sequence variants of a gene ([SO:0001023](http://www.sequenceontology.org/browser/current_svn/term/SO:0001023)). Nucleotide(s)/amino acids from start position of sequence to stop position of sequence on the positive (+) strand of the reference sequence. When the sequence  type is DNA, it should be the sequence on the positive (+) strand. This will lay in the range between variant.start and variant.end.
   */
  referenceAllele?: string|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.variant.referenceAllele
   */
  _referenceAllele?: fhir.IFhirElement|undefined;
  /**
   * Extended CIGAR string for aligning the sequence with reference bases. See detailed documentation [here](http://support.illumina.com/help/SequencingAnalysisWorkflow/Content/Vault/Informatics/Sequencing_Analysis/CASAVA/swSEQ_mCA_ExtendedCIGARFormat.htm).
   */
  cigar?: string|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.variant.cigar
   */
  _cigar?: fhir.IFhirElement|undefined;
  /**
   * A pointer to an Observation containing variant information.
   */
  variantPointer?: fhir.IReference|undefined;
}

/**
 * Receiver Operator Characteristic (ROC) Curve  to give sensitivity/specificity tradeoff.
 */
export type IMolecularSequenceQualityRoc = fhir.IBackboneElement & { 
  /**
   * Invidual data point representing the GQ (genotype quality) score threshold.
   */
  score?: number[]|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.roc.score
   */
  _score?: fhir.IFhirElement[]|undefined;
  /**
   * The number of true positives if the GQ score threshold was set to "score" field value.
   */
  numTP?: number[]|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.roc.numTP
   */
  _numTP?: fhir.IFhirElement[]|undefined;
  /**
   * The number of false positives if the GQ score threshold was set to "score" field value.
   */
  numFP?: number[]|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.roc.numFP
   */
  _numFP?: fhir.IFhirElement[]|undefined;
  /**
   * The number of false negatives if the GQ score threshold was set to "score" field value.
   */
  numFN?: number[]|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.roc.numFN
   */
  _numFN?: fhir.IFhirElement[]|undefined;
  /**
   * Calculated precision if the GQ score threshold was set to "score" field value.
   */
  precision?: number[]|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.roc.precision
   */
  _precision?: fhir.IFhirElement[]|undefined;
  /**
   * Calculated sensitivity if the GQ score threshold was set to "score" field value.
   */
  sensitivity?: number[]|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.roc.sensitivity
   */
  _sensitivity?: fhir.IFhirElement[]|undefined;
  /**
   * Calculated fScore if the GQ score threshold was set to "score" field value.
   */
  fMeasure?: number[]|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.roc.fMeasure
   */
  _fMeasure?: fhir.IFhirElement[]|undefined;
}

/**
 * An experimental feature attribute that defines the quality of the feature in a quantitative way, such as a phred quality score ([SO:0001686](http://www.sequenceontology.org/browser/current_svn/term/SO:0001686)).
 */
export type IMolecularSequenceQuality = fhir.IBackboneElement & { 
  /**
   * INDEL / SNP / Undefined variant.
   */
  type: QualityTypeValueSetEnum|null;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.type
   */
  _type?: fhir.IFhirElement|undefined;
  /**
   * Gold standard sequence used for comparing against.
   */
  standardSequence?: fhir.ICodeableConcept|undefined;
  /**
   * Start position of the sequence. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  start?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.start
   */
  _start?: fhir.IFhirElement|undefined;
  /**
   * End position of the sequence. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  end?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.end
   */
  _end?: fhir.IFhirElement|undefined;
  /**
   * The score of an experimentally derived feature such as a p-value ([SO:0001685](http://www.sequenceontology.org/browser/current_svn/term/SO:0001685)).
   */
  score?: fhir.IQuantity|undefined;
  /**
   * Which method is used to get sequence quality.
   */
  method?: fhir.ICodeableConcept|undefined;
  /**
   * True positives, from the perspective of the truth data, i.e. the number of sites in the Truth Call Set for which there are paths through the Query Call Set that are consistent with all of the alleles at this site, and for which there is an accurate genotype call for the event.
   */
  truthTP?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.truthTP
   */
  _truthTP?: fhir.IFhirElement|undefined;
  /**
   * True positives, from the perspective of the query data, i.e. the number of sites in the Query Call Set for which there are paths through the Truth Call Set that are consistent with all of the alleles at this site, and for which there is an accurate genotype call for the event.
   */
  queryTP?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.queryTP
   */
  _queryTP?: fhir.IFhirElement|undefined;
  /**
   * False negatives, i.e. the number of sites in the Truth Call Set for which there is no path through the Query Call Set that is consistent with all of the alleles at this site, or sites for which there is an inaccurate genotype call for the event. Sites with correct variant but incorrect genotype are counted here.
   */
  truthFN?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.truthFN
   */
  _truthFN?: fhir.IFhirElement|undefined;
  /**
   * False positives, i.e. the number of sites in the Query Call Set for which there is no path through the Truth Call Set that is consistent with this site. Sites with correct variant but incorrect genotype are counted here.
   */
  queryFP?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.queryFP
   */
  _queryFP?: fhir.IFhirElement|undefined;
  /**
   * The number of false positives where the non-REF alleles in the Truth and Query Call Sets match (i.e. cases where the truth is 1/1 and the query is 0/1 or similar).
   */
  gtFP?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.gtFP
   */
  _gtFP?: fhir.IFhirElement|undefined;
  /**
   * QUERY.TP / (QUERY.TP + QUERY.FP).
   */
  precision?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.precision
   */
  _precision?: fhir.IFhirElement|undefined;
  /**
   * TRUTH.TP / (TRUTH.TP + TRUTH.FN).
   */
  recall?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.recall
   */
  _recall?: fhir.IFhirElement|undefined;
  /**
   * Harmonic mean of Recall and Precision, computed as: 2 * precision * recall / (precision + recall).
   */
  fScore?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.fScore
   */
  _fScore?: fhir.IFhirElement|undefined;
  /**
   * Receiver Operator Characteristic (ROC) Curve  to give sensitivity/specificity tradeoff.
   */
  roc?: fhir.IMolecularSequenceQualityRoc|undefined;
}

/**
 * Configurations of the external repository. The repository shall store target's observedSeq or records related with target's observedSeq.
 */
export type IMolecularSequenceRepository = fhir.IBackboneElement & { 
  /**
   * Click and see / RESTful API / Need login to see / RESTful API with authentication / Other ways to see resource.
   */
  type: RepositoryTypeValueSetEnum|null;
  /**
   * Extended properties for primitive element: MolecularSequence.repository.type
   */
  _type?: fhir.IFhirElement|undefined;
  /**
   * URI of an external repository which contains further details about the genetics data.
   */
  url?: string|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.repository.url
   */
  _url?: fhir.IFhirElement|undefined;
  /**
   * URI of an external repository which contains further details about the genetics data.
   */
  name?: string|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.repository.name
   */
  _name?: fhir.IFhirElement|undefined;
  /**
   * Id of the variant in this external repository. The server will understand how to use this id to call for more info about datasets in external repository.
   */
  datasetId?: string|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.repository.datasetId
   */
  _datasetId?: fhir.IFhirElement|undefined;
  /**
   * Id of the variantset in this external repository. The server will understand how to use this id to call for more info about variantsets in external repository.
   */
  variantsetId?: string|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.repository.variantsetId
   */
  _variantsetId?: fhir.IFhirElement|undefined;
  /**
   * Id of the read in this external repository.
   */
  readsetId?: string|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.repository.readsetId
   */
  _readsetId?: fhir.IFhirElement|undefined;
}

/**
 * Structural variant outer.
 */
export type IMolecularSequenceStructureVariantOuter = fhir.IBackboneElement & { 
  /**
   * Structural variant outer start. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  start?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.structureVariant.outer.start
   */
  _start?: fhir.IFhirElement|undefined;
  /**
   * Structural variant outer end. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  end?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.structureVariant.outer.end
   */
  _end?: fhir.IFhirElement|undefined;
}

/**
 * Structural variant inner.
 */
export type IMolecularSequenceStructureVariantInner = fhir.IBackboneElement & { 
  /**
   * Structural variant inner start. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  start?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.structureVariant.inner.start
   */
  _start?: fhir.IFhirElement|undefined;
  /**
   * Structural variant inner end. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  end?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.structureVariant.inner.end
   */
  _end?: fhir.IFhirElement|undefined;
}

/**
 * Information about chromosome structure variation.
 */
export type IMolecularSequenceStructureVariant = fhir.IBackboneElement & { 
  /**
   * Information about chromosome structure variation DNA change type.
   */
  variantType?: fhir.ICodeableConcept|undefined;
  /**
   * Used to indicate if the outer and inner start-end values have the same meaning.
   */
  exact?: boolean|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.structureVariant.exact
   */
  _exact?: fhir.IFhirElement|undefined;
  /**
   * Length of the variant chromosome.
   */
  length?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.structureVariant.length
   */
  _length?: fhir.IFhirElement|undefined;
  /**
   * Structural variant outer.
   */
  outer?: fhir.IMolecularSequenceStructureVariantOuter|undefined;
  /**
   * Structural variant inner.
   */
  inner?: fhir.IMolecularSequenceStructureVariantInner|undefined;
}

/**
 * Raw data describing a biological sequence.
 */
export type IMolecularSequence = fhir.IDomainResource & { 
  /**
   * Resource Type Name
   */
  resourceType: "MolecularSequence";
  /**
   * A unique identifier for this particular sequence instance. This is a FHIR-defined id.
   */
  identifier?: fhir.IIdentifier[]|undefined;
  /**
   * Amino Acid Sequence/ DNA Sequence / RNA Sequence.
   */
  type?: SequenceTypeValueSetEnum|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.type
   */
  _type?: fhir.IFhirElement|undefined;
  /**
   * Whether the sequence is numbered starting at 0 (0-based numbering or coordinates, inclusive start, exclusive end) or starting at 1 (1-based numbering, inclusive start and inclusive end).
   */
  coordinateSystem: number|null;
  /**
   * Extended properties for primitive element: MolecularSequence.coordinateSystem
   */
  _coordinateSystem?: fhir.IFhirElement|undefined;
  /**
   * The patient whose sequencing results are described by this resource.
   */
  patient?: fhir.IReference|undefined;
  /**
   * Specimen used for sequencing.
   */
  specimen?: fhir.IReference|undefined;
  /**
   * The method for sequencing, for example, chip information.
   */
  device?: fhir.IReference|undefined;
  /**
   * The organization or lab that should be responsible for this result.
   */
  performer?: fhir.IReference|undefined;
  /**
   * The number of copies of the sequence of interest. (RNASeq).
   */
  quantity?: fhir.IQuantity|undefined;
  /**
   * A sequence that is used as a reference to describe variants that are present in a sequence analyzed.
   */
  referenceSeq?: fhir.IMolecularSequenceReferenceSeq|undefined;
  /**
   * The definition of variant here originates from Sequence ontology ([variant_of](http://www.sequenceontology.org/browser/current_svn/term/variant_of)). This element can represent amino acid or nucleic sequence change(including insertion,deletion,SNP,etc.)  It can represent some complex mutation or segment variation with the assist of CIGAR string.
   */
  variant?: fhir.IMolecularSequenceVariant[]|undefined;
  /**
   * Sequence that was observed. It is the result marked by referenceSeq along with variant records on referenceSeq. This shall start from referenceSeq.windowStart and end by referenceSeq.windowEnd.
   */
  observedSeq?: string|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.observedSeq
   */
  _observedSeq?: fhir.IFhirElement|undefined;
  /**
   * An experimental feature attribute that defines the quality of the feature in a quantitative way, such as a phred quality score ([SO:0001686](http://www.sequenceontology.org/browser/current_svn/term/SO:0001686)).
   */
  quality?: fhir.IMolecularSequenceQuality[]|undefined;
  /**
   * Coverage (read depth or depth) is the average number of reads representing a given nucleotide in the reconstructed sequence.
   */
  readCoverage?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.readCoverage
   */
  _readCoverage?: fhir.IFhirElement|undefined;
  /**
   * Configurations of the external repository. The repository shall store target's observedSeq or records related with target's observedSeq.
   */
  repository?: fhir.IMolecularSequenceRepository[]|undefined;
  /**
   * Pointer to next atomic sequence which at most contains one variant.
   */
  pointer?: fhir.IReference[]|undefined;
  /**
   * Information about chromosome structure variation.
   */
  structureVariant?: fhir.IMolecularSequenceStructureVariant[]|undefined;
}

/**
 * A sequence that is used as a reference to describe variants that are present in a sequence analyzed.
 */
export class MolecularSequenceReferenceSeq extends fhir.BackboneElement implements IMolecularSequenceReferenceSeq {
  /**
   * Structural unit composed of a nucleic acid molecule which controls its own replication through the interaction of specific proteins at one or more origins of replication ([SO:0000340](http://www.sequenceontology.org/browser/current_svn/term/SO:0000340)).
   */
  public chromosome?: fhir.CodeableConcept|undefined;
  /**
   * The Genome Build used for reference, following GRCh build versions e.g. 'GRCh 37'.  Version number must be included if a versioned release of a primary build was used.
   */
  public genomeBuild?: string|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.referenceSeq.genomeBuild
   */
  public _genomeBuild?: fhir.FhirElement|undefined;
  /**
   * A relative reference to a DNA strand based on gene orientation. The strand that contains the open reading frame of the gene is the "sense" strand, and the opposite complementary strand is the "antisense" strand.
   */
  public orientation?: OrientationTypeValueSetEnum|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.referenceSeq.orientation
   */
  public _orientation?: fhir.FhirElement|undefined;
  /**
   * Reference identifier of reference sequence submitted to NCBI. It must match the type in the MolecularSequence.type field. For example, the prefix, “NG_” identifies reference sequence for genes, “NM_” for messenger RNA transcripts, and “NP_” for amino acid sequences.
   */
  public referenceSeqId?: fhir.CodeableConcept|undefined;
  /**
   * A pointer to another MolecularSequence entity as reference sequence.
   */
  public referenceSeqPointer?: fhir.Reference|undefined;
  /**
   * A string like "ACGT".
   */
  public referenceSeqString?: string|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.referenceSeq.referenceSeqString
   */
  public _referenceSeqString?: fhir.FhirElement|undefined;
  /**
   * An absolute reference to a strand. The Watson strand is the strand whose 5'-end is on the short arm of the chromosome, and the Crick strand as the one whose 5'-end is on the long arm.
   */
  public strand?: StrandTypeValueSetEnum|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.referenceSeq.strand
   */
  public _strand?: fhir.FhirElement|undefined;
  /**
   * Start position of the window on the reference sequence. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  public windowStart?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.referenceSeq.windowStart
   */
  public _windowStart?: fhir.FhirElement|undefined;
  /**
   * End position of the window on the reference sequence. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  public windowEnd?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.referenceSeq.windowEnd
   */
  public _windowEnd?: fhir.FhirElement|undefined;
  /**
   * Default constructor for MolecularSequenceReferenceSeq - initializes any required elements to null if a value is not provided.
   */
  constructor(source:Partial<IMolecularSequenceReferenceSeq> = { }) {
    super(source);
    if (source['chromosome']) { this.chromosome = new fhir.CodeableConcept(source.chromosome!); }
    if (source['genomeBuild']) { this.genomeBuild = source.genomeBuild; }
    if (source['_genomeBuild']) { this._genomeBuild = new fhir.FhirElement(source._genomeBuild!); }
    if (source['orientation']) { this.orientation = source.orientation; }
    if (source['_orientation']) { this._orientation = new fhir.FhirElement(source._orientation!); }
    if (source['referenceSeqId']) { this.referenceSeqId = new fhir.CodeableConcept(source.referenceSeqId!); }
    if (source['referenceSeqPointer']) { this.referenceSeqPointer = new fhir.Reference(source.referenceSeqPointer!); }
    if (source['referenceSeqString']) { this.referenceSeqString = source.referenceSeqString; }
    if (source['_referenceSeqString']) { this._referenceSeqString = new fhir.FhirElement(source._referenceSeqString!); }
    if (source['strand']) { this.strand = source.strand; }
    if (source['_strand']) { this._strand = new fhir.FhirElement(source._strand!); }
    if (source['windowStart']) { this.windowStart = source.windowStart; }
    if (source['_windowStart']) { this._windowStart = new fhir.FhirElement(source._windowStart!); }
    if (source['windowEnd']) { this.windowEnd = source.windowEnd; }
    if (source['_windowEnd']) { this._windowEnd = new fhir.FhirElement(source._windowEnd!); }
  }
  /**
   * Example-bound Value Set for chromosome
   */
  public static chromosomeExampleValueSet():ChromosomeHumanValueSetType {
    return ChromosomeHumanValueSet;
  }
  /**
   * Required-bound Value Set for orientation
   */
  public static orientationRequiredValueSet():OrientationTypeValueSetType {
    return OrientationTypeValueSet;
  }
  /**
   * Required-bound Value Set for strand
   */
  public static strandRequiredValueSet():StrandTypeValueSetType {
    return StrandTypeValueSet;
  }
  /**
   * Function to perform basic model validation (e.g., check if required elements are present).
   */
  public override doModelValidation():[string,string][] {
    var results:[string,string][] = super.doModelValidation();
    if (this["chromosome"]) { results.push(...this.chromosome.doModelValidation()); }
    if (this["_genomeBuild"]) { results.push(...this._genomeBuild.doModelValidation()); }
    if (this["_orientation"]) { results.push(...this._orientation.doModelValidation()); }
    if (this["referenceSeqId"]) { results.push(...this.referenceSeqId.doModelValidation()); }
    if (this["referenceSeqPointer"]) { results.push(...this.referenceSeqPointer.doModelValidation()); }
    if (this["_referenceSeqString"]) { results.push(...this._referenceSeqString.doModelValidation()); }
    if (this["_strand"]) { results.push(...this._strand.doModelValidation()); }
    if (this["_windowStart"]) { results.push(...this._windowStart.doModelValidation()); }
    if (this["_windowEnd"]) { results.push(...this._windowEnd.doModelValidation()); }
    return results;
  }
}

/**
 * The definition of variant here originates from Sequence ontology ([variant_of](http://www.sequenceontology.org/browser/current_svn/term/variant_of)). This element can represent amino acid or nucleic sequence change(including insertion,deletion,SNP,etc.)  It can represent some complex mutation or segment variation with the assist of CIGAR string.
 */
export class MolecularSequenceVariant extends fhir.BackboneElement implements IMolecularSequenceVariant {
  /**
   * Start position of the variant on the  reference sequence. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  public start?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.variant.start
   */
  public _start?: fhir.FhirElement|undefined;
  /**
   * End position of the variant on the reference sequence. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  public end?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.variant.end
   */
  public _end?: fhir.FhirElement|undefined;
  /**
   * An allele is one of a set of coexisting sequence variants of a gene ([SO:0001023](http://www.sequenceontology.org/browser/current_svn/term/SO:0001023)).  Nucleotide(s)/amino acids from start position of sequence to stop position of sequence on the positive (+) strand of the observed  sequence. When the sequence  type is DNA, it should be the sequence on the positive (+) strand. This will lay in the range between variant.start and variant.end.
   */
  public observedAllele?: string|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.variant.observedAllele
   */
  public _observedAllele?: fhir.FhirElement|undefined;
  /**
   * An allele is one of a set of coexisting sequence variants of a gene ([SO:0001023](http://www.sequenceontology.org/browser/current_svn/term/SO:0001023)). Nucleotide(s)/amino acids from start position of sequence to stop position of sequence on the positive (+) strand of the reference sequence. When the sequence  type is DNA, it should be the sequence on the positive (+) strand. This will lay in the range between variant.start and variant.end.
   */
  public referenceAllele?: string|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.variant.referenceAllele
   */
  public _referenceAllele?: fhir.FhirElement|undefined;
  /**
   * Extended CIGAR string for aligning the sequence with reference bases. See detailed documentation [here](http://support.illumina.com/help/SequencingAnalysisWorkflow/Content/Vault/Informatics/Sequencing_Analysis/CASAVA/swSEQ_mCA_ExtendedCIGARFormat.htm).
   */
  public cigar?: string|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.variant.cigar
   */
  public _cigar?: fhir.FhirElement|undefined;
  /**
   * A pointer to an Observation containing variant information.
   */
  public variantPointer?: fhir.Reference|undefined;
  /**
   * Default constructor for MolecularSequenceVariant - initializes any required elements to null if a value is not provided.
   */
  constructor(source:Partial<IMolecularSequenceVariant> = { }) {
    super(source);
    if (source['start']) { this.start = source.start; }
    if (source['_start']) { this._start = new fhir.FhirElement(source._start!); }
    if (source['end']) { this.end = source.end; }
    if (source['_end']) { this._end = new fhir.FhirElement(source._end!); }
    if (source['observedAllele']) { this.observedAllele = source.observedAllele; }
    if (source['_observedAllele']) { this._observedAllele = new fhir.FhirElement(source._observedAllele!); }
    if (source['referenceAllele']) { this.referenceAllele = source.referenceAllele; }
    if (source['_referenceAllele']) { this._referenceAllele = new fhir.FhirElement(source._referenceAllele!); }
    if (source['cigar']) { this.cigar = source.cigar; }
    if (source['_cigar']) { this._cigar = new fhir.FhirElement(source._cigar!); }
    if (source['variantPointer']) { this.variantPointer = new fhir.Reference(source.variantPointer!); }
  }
  /**
   * Function to perform basic model validation (e.g., check if required elements are present).
   */
  public override doModelValidation():[string,string][] {
    var results:[string,string][] = super.doModelValidation();
    if (this["_start"]) { results.push(...this._start.doModelValidation()); }
    if (this["_end"]) { results.push(...this._end.doModelValidation()); }
    if (this["_observedAllele"]) { results.push(...this._observedAllele.doModelValidation()); }
    if (this["_referenceAllele"]) { results.push(...this._referenceAllele.doModelValidation()); }
    if (this["_cigar"]) { results.push(...this._cigar.doModelValidation()); }
    if (this["variantPointer"]) { results.push(...this.variantPointer.doModelValidation()); }
    return results;
  }
}

/**
 * Receiver Operator Characteristic (ROC) Curve  to give sensitivity/specificity tradeoff.
 */
export class MolecularSequenceQualityRoc extends fhir.BackboneElement implements IMolecularSequenceQualityRoc {
  /**
   * Invidual data point representing the GQ (genotype quality) score threshold.
   */
  public score?: number[]|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.roc.score
   */
  public _score?: fhir.FhirElement[]|undefined;
  /**
   * The number of true positives if the GQ score threshold was set to "score" field value.
   */
  public numTP?: number[]|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.roc.numTP
   */
  public _numTP?: fhir.FhirElement[]|undefined;
  /**
   * The number of false positives if the GQ score threshold was set to "score" field value.
   */
  public numFP?: number[]|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.roc.numFP
   */
  public _numFP?: fhir.FhirElement[]|undefined;
  /**
   * The number of false negatives if the GQ score threshold was set to "score" field value.
   */
  public numFN?: number[]|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.roc.numFN
   */
  public _numFN?: fhir.FhirElement[]|undefined;
  /**
   * Calculated precision if the GQ score threshold was set to "score" field value.
   */
  public precision?: number[]|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.roc.precision
   */
  public _precision?: fhir.FhirElement[]|undefined;
  /**
   * Calculated sensitivity if the GQ score threshold was set to "score" field value.
   */
  public sensitivity?: number[]|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.roc.sensitivity
   */
  public _sensitivity?: fhir.FhirElement[]|undefined;
  /**
   * Calculated fScore if the GQ score threshold was set to "score" field value.
   */
  public fMeasure?: number[]|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.roc.fMeasure
   */
  public _fMeasure?: fhir.FhirElement[]|undefined;
  /**
   * Default constructor for MolecularSequenceQualityRoc - initializes any required elements to null if a value is not provided.
   */
  constructor(source:Partial<IMolecularSequenceQualityRoc> = { }) {
    super(source);
    if (source['score']) { this.score = source.score.map((x) => (x)); }
    if (source['_score']) { this._score = source._score.map((x) => new fhir.FhirElement(x)); }
    if (source['numTP']) { this.numTP = source.numTP.map((x) => (x)); }
    if (source['_numTP']) { this._numTP = source._numTP.map((x) => new fhir.FhirElement(x)); }
    if (source['numFP']) { this.numFP = source.numFP.map((x) => (x)); }
    if (source['_numFP']) { this._numFP = source._numFP.map((x) => new fhir.FhirElement(x)); }
    if (source['numFN']) { this.numFN = source.numFN.map((x) => (x)); }
    if (source['_numFN']) { this._numFN = source._numFN.map((x) => new fhir.FhirElement(x)); }
    if (source['precision']) { this.precision = source.precision.map((x) => (x)); }
    if (source['_precision']) { this._precision = source._precision.map((x) => new fhir.FhirElement(x)); }
    if (source['sensitivity']) { this.sensitivity = source.sensitivity.map((x) => (x)); }
    if (source['_sensitivity']) { this._sensitivity = source._sensitivity.map((x) => new fhir.FhirElement(x)); }
    if (source['fMeasure']) { this.fMeasure = source.fMeasure.map((x) => (x)); }
    if (source['_fMeasure']) { this._fMeasure = source._fMeasure.map((x) => new fhir.FhirElement(x)); }
  }
  /**
   * Function to perform basic model validation (e.g., check if required elements are present).
   */
  public override doModelValidation():[string,string][] {
    var results:[string,string][] = super.doModelValidation();
    if (this["_score"]) { this._score.forEach((x) => { results.push(...x.doModelValidation()); }) }
    if (this["_numTP"]) { this._numTP.forEach((x) => { results.push(...x.doModelValidation()); }) }
    if (this["_numFP"]) { this._numFP.forEach((x) => { results.push(...x.doModelValidation()); }) }
    if (this["_numFN"]) { this._numFN.forEach((x) => { results.push(...x.doModelValidation()); }) }
    if (this["_precision"]) { this._precision.forEach((x) => { results.push(...x.doModelValidation()); }) }
    if (this["_sensitivity"]) { this._sensitivity.forEach((x) => { results.push(...x.doModelValidation()); }) }
    if (this["_fMeasure"]) { this._fMeasure.forEach((x) => { results.push(...x.doModelValidation()); }) }
    return results;
  }
}

/**
 * An experimental feature attribute that defines the quality of the feature in a quantitative way, such as a phred quality score ([SO:0001686](http://www.sequenceontology.org/browser/current_svn/term/SO:0001686)).
 */
export class MolecularSequenceQuality extends fhir.BackboneElement implements IMolecularSequenceQuality {
  /**
   * INDEL / SNP / Undefined variant.
   */
  public type: QualityTypeValueSetEnum|null;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.type
   */
  public _type?: fhir.FhirElement|undefined;
  /**
   * Gold standard sequence used for comparing against.
   */
  public standardSequence?: fhir.CodeableConcept|undefined;
  /**
   * Start position of the sequence. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  public start?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.start
   */
  public _start?: fhir.FhirElement|undefined;
  /**
   * End position of the sequence. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  public end?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.end
   */
  public _end?: fhir.FhirElement|undefined;
  /**
   * The score of an experimentally derived feature such as a p-value ([SO:0001685](http://www.sequenceontology.org/browser/current_svn/term/SO:0001685)).
   */
  public score?: fhir.Quantity|undefined;
  /**
   * Which method is used to get sequence quality.
   */
  public method?: fhir.CodeableConcept|undefined;
  /**
   * True positives, from the perspective of the truth data, i.e. the number of sites in the Truth Call Set for which there are paths through the Query Call Set that are consistent with all of the alleles at this site, and for which there is an accurate genotype call for the event.
   */
  public truthTP?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.truthTP
   */
  public _truthTP?: fhir.FhirElement|undefined;
  /**
   * True positives, from the perspective of the query data, i.e. the number of sites in the Query Call Set for which there are paths through the Truth Call Set that are consistent with all of the alleles at this site, and for which there is an accurate genotype call for the event.
   */
  public queryTP?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.queryTP
   */
  public _queryTP?: fhir.FhirElement|undefined;
  /**
   * False negatives, i.e. the number of sites in the Truth Call Set for which there is no path through the Query Call Set that is consistent with all of the alleles at this site, or sites for which there is an inaccurate genotype call for the event. Sites with correct variant but incorrect genotype are counted here.
   */
  public truthFN?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.truthFN
   */
  public _truthFN?: fhir.FhirElement|undefined;
  /**
   * False positives, i.e. the number of sites in the Query Call Set for which there is no path through the Truth Call Set that is consistent with this site. Sites with correct variant but incorrect genotype are counted here.
   */
  public queryFP?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.queryFP
   */
  public _queryFP?: fhir.FhirElement|undefined;
  /**
   * The number of false positives where the non-REF alleles in the Truth and Query Call Sets match (i.e. cases where the truth is 1/1 and the query is 0/1 or similar).
   */
  public gtFP?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.gtFP
   */
  public _gtFP?: fhir.FhirElement|undefined;
  /**
   * QUERY.TP / (QUERY.TP + QUERY.FP).
   */
  public precision?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.precision
   */
  public _precision?: fhir.FhirElement|undefined;
  /**
   * TRUTH.TP / (TRUTH.TP + TRUTH.FN).
   */
  public recall?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.recall
   */
  public _recall?: fhir.FhirElement|undefined;
  /**
   * Harmonic mean of Recall and Precision, computed as: 2 * precision * recall / (precision + recall).
   */
  public fScore?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.quality.fScore
   */
  public _fScore?: fhir.FhirElement|undefined;
  /**
   * Receiver Operator Characteristic (ROC) Curve  to give sensitivity/specificity tradeoff.
   */
  public roc?: fhir.MolecularSequenceQualityRoc|undefined;
  /**
   * Default constructor for MolecularSequenceQuality - initializes any required elements to null if a value is not provided.
   */
  constructor(source:Partial<IMolecularSequenceQuality> = { }) {
    super(source);
    if (source['type']) { this.type = source.type; }
    else { this.type = null; }
    if (source['_type']) { this._type = new fhir.FhirElement(source._type!); }
    if (source['standardSequence']) { this.standardSequence = new fhir.CodeableConcept(source.standardSequence!); }
    if (source['start']) { this.start = source.start; }
    if (source['_start']) { this._start = new fhir.FhirElement(source._start!); }
    if (source['end']) { this.end = source.end; }
    if (source['_end']) { this._end = new fhir.FhirElement(source._end!); }
    if (source['score']) { this.score = new fhir.Quantity(source.score!); }
    if (source['method']) { this.method = new fhir.CodeableConcept(source.method!); }
    if (source['truthTP']) { this.truthTP = source.truthTP; }
    if (source['_truthTP']) { this._truthTP = new fhir.FhirElement(source._truthTP!); }
    if (source['queryTP']) { this.queryTP = source.queryTP; }
    if (source['_queryTP']) { this._queryTP = new fhir.FhirElement(source._queryTP!); }
    if (source['truthFN']) { this.truthFN = source.truthFN; }
    if (source['_truthFN']) { this._truthFN = new fhir.FhirElement(source._truthFN!); }
    if (source['queryFP']) { this.queryFP = source.queryFP; }
    if (source['_queryFP']) { this._queryFP = new fhir.FhirElement(source._queryFP!); }
    if (source['gtFP']) { this.gtFP = source.gtFP; }
    if (source['_gtFP']) { this._gtFP = new fhir.FhirElement(source._gtFP!); }
    if (source['precision']) { this.precision = source.precision; }
    if (source['_precision']) { this._precision = new fhir.FhirElement(source._precision!); }
    if (source['recall']) { this.recall = source.recall; }
    if (source['_recall']) { this._recall = new fhir.FhirElement(source._recall!); }
    if (source['fScore']) { this.fScore = source.fScore; }
    if (source['_fScore']) { this._fScore = new fhir.FhirElement(source._fScore!); }
    if (source['roc']) { this.roc = new fhir.MolecularSequenceQualityRoc(source.roc!); }
  }
  /**
   * Required-bound Value Set for type
   */
  public static typeRequiredValueSet():QualityTypeValueSetType {
    return QualityTypeValueSet;
  }
  /**
   * Function to perform basic model validation (e.g., check if required elements are present).
   */
  public override doModelValidation():[string,string][] {
    var results:[string,string][] = super.doModelValidation();
    if (!this["type"]) { results.push(["type",'Missing required element: MolecularSequence.quality.type']); }
    if (this["_type"]) { results.push(...this._type.doModelValidation()); }
    if (this["standardSequence"]) { results.push(...this.standardSequence.doModelValidation()); }
    if (this["_start"]) { results.push(...this._start.doModelValidation()); }
    if (this["_end"]) { results.push(...this._end.doModelValidation()); }
    if (this["score"]) { results.push(...this.score.doModelValidation()); }
    if (this["method"]) { results.push(...this.method.doModelValidation()); }
    if (this["_truthTP"]) { results.push(...this._truthTP.doModelValidation()); }
    if (this["_queryTP"]) { results.push(...this._queryTP.doModelValidation()); }
    if (this["_truthFN"]) { results.push(...this._truthFN.doModelValidation()); }
    if (this["_queryFP"]) { results.push(...this._queryFP.doModelValidation()); }
    if (this["_gtFP"]) { results.push(...this._gtFP.doModelValidation()); }
    if (this["_precision"]) { results.push(...this._precision.doModelValidation()); }
    if (this["_recall"]) { results.push(...this._recall.doModelValidation()); }
    if (this["_fScore"]) { results.push(...this._fScore.doModelValidation()); }
    if (this["roc"]) { results.push(...this.roc.doModelValidation()); }
    return results;
  }
}

/**
 * Configurations of the external repository. The repository shall store target's observedSeq or records related with target's observedSeq.
 */
export class MolecularSequenceRepository extends fhir.BackboneElement implements IMolecularSequenceRepository {
  /**
   * Click and see / RESTful API / Need login to see / RESTful API with authentication / Other ways to see resource.
   */
  public type: RepositoryTypeValueSetEnum|null;
  /**
   * Extended properties for primitive element: MolecularSequence.repository.type
   */
  public _type?: fhir.FhirElement|undefined;
  /**
   * URI of an external repository which contains further details about the genetics data.
   */
  public url?: string|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.repository.url
   */
  public _url?: fhir.FhirElement|undefined;
  /**
   * URI of an external repository which contains further details about the genetics data.
   */
  public name?: string|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.repository.name
   */
  public _name?: fhir.FhirElement|undefined;
  /**
   * Id of the variant in this external repository. The server will understand how to use this id to call for more info about datasets in external repository.
   */
  public datasetId?: string|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.repository.datasetId
   */
  public _datasetId?: fhir.FhirElement|undefined;
  /**
   * Id of the variantset in this external repository. The server will understand how to use this id to call for more info about variantsets in external repository.
   */
  public variantsetId?: string|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.repository.variantsetId
   */
  public _variantsetId?: fhir.FhirElement|undefined;
  /**
   * Id of the read in this external repository.
   */
  public readsetId?: string|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.repository.readsetId
   */
  public _readsetId?: fhir.FhirElement|undefined;
  /**
   * Default constructor for MolecularSequenceRepository - initializes any required elements to null if a value is not provided.
   */
  constructor(source:Partial<IMolecularSequenceRepository> = { }) {
    super(source);
    if (source['type']) { this.type = source.type; }
    else { this.type = null; }
    if (source['_type']) { this._type = new fhir.FhirElement(source._type!); }
    if (source['url']) { this.url = source.url; }
    if (source['_url']) { this._url = new fhir.FhirElement(source._url!); }
    if (source['name']) { this.name = source.name; }
    if (source['_name']) { this._name = new fhir.FhirElement(source._name!); }
    if (source['datasetId']) { this.datasetId = source.datasetId; }
    if (source['_datasetId']) { this._datasetId = new fhir.FhirElement(source._datasetId!); }
    if (source['variantsetId']) { this.variantsetId = source.variantsetId; }
    if (source['_variantsetId']) { this._variantsetId = new fhir.FhirElement(source._variantsetId!); }
    if (source['readsetId']) { this.readsetId = source.readsetId; }
    if (source['_readsetId']) { this._readsetId = new fhir.FhirElement(source._readsetId!); }
  }
  /**
   * Required-bound Value Set for type
   */
  public static typeRequiredValueSet():RepositoryTypeValueSetType {
    return RepositoryTypeValueSet;
  }
  /**
   * Function to perform basic model validation (e.g., check if required elements are present).
   */
  public override doModelValidation():[string,string][] {
    var results:[string,string][] = super.doModelValidation();
    if (!this["type"]) { results.push(["type",'Missing required element: MolecularSequence.repository.type']); }
    if (this["_type"]) { results.push(...this._type.doModelValidation()); }
    if (this["_url"]) { results.push(...this._url.doModelValidation()); }
    if (this["_name"]) { results.push(...this._name.doModelValidation()); }
    if (this["_datasetId"]) { results.push(...this._datasetId.doModelValidation()); }
    if (this["_variantsetId"]) { results.push(...this._variantsetId.doModelValidation()); }
    if (this["_readsetId"]) { results.push(...this._readsetId.doModelValidation()); }
    return results;
  }
}

/**
 * Structural variant outer.
 */
export class MolecularSequenceStructureVariantOuter extends fhir.BackboneElement implements IMolecularSequenceStructureVariantOuter {
  /**
   * Structural variant outer start. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  public start?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.structureVariant.outer.start
   */
  public _start?: fhir.FhirElement|undefined;
  /**
   * Structural variant outer end. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  public end?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.structureVariant.outer.end
   */
  public _end?: fhir.FhirElement|undefined;
  /**
   * Default constructor for MolecularSequenceStructureVariantOuter - initializes any required elements to null if a value is not provided.
   */
  constructor(source:Partial<IMolecularSequenceStructureVariantOuter> = { }) {
    super(source);
    if (source['start']) { this.start = source.start; }
    if (source['_start']) { this._start = new fhir.FhirElement(source._start!); }
    if (source['end']) { this.end = source.end; }
    if (source['_end']) { this._end = new fhir.FhirElement(source._end!); }
  }
  /**
   * Function to perform basic model validation (e.g., check if required elements are present).
   */
  public override doModelValidation():[string,string][] {
    var results:[string,string][] = super.doModelValidation();
    if (this["_start"]) { results.push(...this._start.doModelValidation()); }
    if (this["_end"]) { results.push(...this._end.doModelValidation()); }
    return results;
  }
}

/**
 * Structural variant inner.
 */
export class MolecularSequenceStructureVariantInner extends fhir.BackboneElement implements IMolecularSequenceStructureVariantInner {
  /**
   * Structural variant inner start. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  public start?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.structureVariant.inner.start
   */
  public _start?: fhir.FhirElement|undefined;
  /**
   * Structural variant inner end. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  public end?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.structureVariant.inner.end
   */
  public _end?: fhir.FhirElement|undefined;
  /**
   * Default constructor for MolecularSequenceStructureVariantInner - initializes any required elements to null if a value is not provided.
   */
  constructor(source:Partial<IMolecularSequenceStructureVariantInner> = { }) {
    super(source);
    if (source['start']) { this.start = source.start; }
    if (source['_start']) { this._start = new fhir.FhirElement(source._start!); }
    if (source['end']) { this.end = source.end; }
    if (source['_end']) { this._end = new fhir.FhirElement(source._end!); }
  }
  /**
   * Function to perform basic model validation (e.g., check if required elements are present).
   */
  public override doModelValidation():[string,string][] {
    var results:[string,string][] = super.doModelValidation();
    if (this["_start"]) { results.push(...this._start.doModelValidation()); }
    if (this["_end"]) { results.push(...this._end.doModelValidation()); }
    return results;
  }
}

/**
 * Information about chromosome structure variation.
 */
export class MolecularSequenceStructureVariant extends fhir.BackboneElement implements IMolecularSequenceStructureVariant {
  /**
   * Information about chromosome structure variation DNA change type.
   */
  public variantType?: fhir.CodeableConcept|undefined;
  /**
   * Used to indicate if the outer and inner start-end values have the same meaning.
   */
  public exact?: boolean|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.structureVariant.exact
   */
  public _exact?: fhir.FhirElement|undefined;
  /**
   * Length of the variant chromosome.
   */
  public length?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.structureVariant.length
   */
  public _length?: fhir.FhirElement|undefined;
  /**
   * Structural variant outer.
   */
  public outer?: fhir.MolecularSequenceStructureVariantOuter|undefined;
  /**
   * Structural variant inner.
   */
  public inner?: fhir.MolecularSequenceStructureVariantInner|undefined;
  /**
   * Default constructor for MolecularSequenceStructureVariant - initializes any required elements to null if a value is not provided.
   */
  constructor(source:Partial<IMolecularSequenceStructureVariant> = { }) {
    super(source);
    if (source['variantType']) { this.variantType = new fhir.CodeableConcept(source.variantType!); }
    if (source['exact']) { this.exact = source.exact; }
    if (source['_exact']) { this._exact = new fhir.FhirElement(source._exact!); }
    if (source['length']) { this.length = source.length; }
    if (source['_length']) { this._length = new fhir.FhirElement(source._length!); }
    if (source['outer']) { this.outer = new fhir.MolecularSequenceStructureVariantOuter(source.outer!); }
    if (source['inner']) { this.inner = new fhir.MolecularSequenceStructureVariantInner(source.inner!); }
  }
  /**
   * Function to perform basic model validation (e.g., check if required elements are present).
   */
  public override doModelValidation():[string,string][] {
    var results:[string,string][] = super.doModelValidation();
    if (this["variantType"]) { results.push(...this.variantType.doModelValidation()); }
    if (this["_exact"]) { results.push(...this._exact.doModelValidation()); }
    if (this["_length"]) { results.push(...this._length.doModelValidation()); }
    if (this["outer"]) { results.push(...this.outer.doModelValidation()); }
    if (this["inner"]) { results.push(...this.inner.doModelValidation()); }
    return results;
  }
}

/**
 * Raw data describing a biological sequence.
 */
export class MolecularSequence extends fhir.DomainResource implements IMolecularSequence {
  /**
   * Resource Type Name
   */
  public resourceType: "MolecularSequence";
  /**
   * A unique identifier for this particular sequence instance. This is a FHIR-defined id.
   */
  public identifier?: fhir.Identifier[]|undefined;
  /**
   * Amino Acid Sequence/ DNA Sequence / RNA Sequence.
   */
  public type?: SequenceTypeValueSetEnum|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.type
   */
  public _type?: fhir.FhirElement|undefined;
  /**
   * Whether the sequence is numbered starting at 0 (0-based numbering or coordinates, inclusive start, exclusive end) or starting at 1 (1-based numbering, inclusive start and inclusive end).
   */
  public coordinateSystem: number|null;
  /**
   * Extended properties for primitive element: MolecularSequence.coordinateSystem
   */
  public _coordinateSystem?: fhir.FhirElement|undefined;
  /**
   * The patient whose sequencing results are described by this resource.
   */
  public patient?: fhir.Reference|undefined;
  /**
   * Specimen used for sequencing.
   */
  public specimen?: fhir.Reference|undefined;
  /**
   * The method for sequencing, for example, chip information.
   */
  public device?: fhir.Reference|undefined;
  /**
   * The organization or lab that should be responsible for this result.
   */
  public performer?: fhir.Reference|undefined;
  /**
   * The number of copies of the sequence of interest. (RNASeq).
   */
  public quantity?: fhir.Quantity|undefined;
  /**
   * A sequence that is used as a reference to describe variants that are present in a sequence analyzed.
   */
  public referenceSeq?: fhir.MolecularSequenceReferenceSeq|undefined;
  /**
   * The definition of variant here originates from Sequence ontology ([variant_of](http://www.sequenceontology.org/browser/current_svn/term/variant_of)). This element can represent amino acid or nucleic sequence change(including insertion,deletion,SNP,etc.)  It can represent some complex mutation or segment variation with the assist of CIGAR string.
   */
  public variant?: fhir.MolecularSequenceVariant[]|undefined;
  /**
   * Sequence that was observed. It is the result marked by referenceSeq along with variant records on referenceSeq. This shall start from referenceSeq.windowStart and end by referenceSeq.windowEnd.
   */
  public observedSeq?: string|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.observedSeq
   */
  public _observedSeq?: fhir.FhirElement|undefined;
  /**
   * An experimental feature attribute that defines the quality of the feature in a quantitative way, such as a phred quality score ([SO:0001686](http://www.sequenceontology.org/browser/current_svn/term/SO:0001686)).
   */
  public quality?: fhir.MolecularSequenceQuality[]|undefined;
  /**
   * Coverage (read depth or depth) is the average number of reads representing a given nucleotide in the reconstructed sequence.
   */
  public readCoverage?: number|undefined;
  /**
   * Extended properties for primitive element: MolecularSequence.readCoverage
   */
  public _readCoverage?: fhir.FhirElement|undefined;
  /**
   * Configurations of the external repository. The repository shall store target's observedSeq or records related with target's observedSeq.
   */
  public repository?: fhir.MolecularSequenceRepository[]|undefined;
  /**
   * Pointer to next atomic sequence which at most contains one variant.
   */
  public pointer?: fhir.Reference[]|undefined;
  /**
   * Information about chromosome structure variation.
   */
  public structureVariant?: fhir.MolecularSequenceStructureVariant[]|undefined;
  /**
   * Default constructor for MolecularSequence - initializes any required elements to null if a value is not provided.
   */
  constructor(source:Partial<IMolecularSequence> = { }) {
    super(source);
    this.resourceType = 'MolecularSequence';
    if (source['identifier']) { this.identifier = source.identifier.map((x) => new fhir.Identifier(x)); }
    if (source['type']) { this.type = source.type; }
    if (source['_type']) { this._type = new fhir.FhirElement(source._type!); }
    if (source['coordinateSystem']) { this.coordinateSystem = source.coordinateSystem; }
    else { this.coordinateSystem = null; }
    if (source['_coordinateSystem']) { this._coordinateSystem = new fhir.FhirElement(source._coordinateSystem!); }
    if (source['patient']) { this.patient = new fhir.Reference(source.patient!); }
    if (source['specimen']) { this.specimen = new fhir.Reference(source.specimen!); }
    if (source['device']) { this.device = new fhir.Reference(source.device!); }
    if (source['performer']) { this.performer = new fhir.Reference(source.performer!); }
    if (source['quantity']) { this.quantity = new fhir.Quantity(source.quantity!); }
    if (source['referenceSeq']) { this.referenceSeq = new fhir.MolecularSequenceReferenceSeq(source.referenceSeq!); }
    if (source['variant']) { this.variant = source.variant.map((x) => new fhir.MolecularSequenceVariant(x)); }
    if (source['observedSeq']) { this.observedSeq = source.observedSeq; }
    if (source['_observedSeq']) { this._observedSeq = new fhir.FhirElement(source._observedSeq!); }
    if (source['quality']) { this.quality = source.quality.map((x) => new fhir.MolecularSequenceQuality(x)); }
    if (source['readCoverage']) { this.readCoverage = source.readCoverage; }
    if (source['_readCoverage']) { this._readCoverage = new fhir.FhirElement(source._readCoverage!); }
    if (source['repository']) { this.repository = source.repository.map((x) => new fhir.MolecularSequenceRepository(x)); }
    if (source['pointer']) { this.pointer = source.pointer.map((x) => new fhir.Reference(x)); }
    if (source['structureVariant']) { this.structureVariant = source.structureVariant.map((x) => new fhir.MolecularSequenceStructureVariant(x)); }
  }
  /**
   * Required-bound Value Set for type
   */
  public static typeRequiredValueSet():SequenceTypeValueSetType {
    return SequenceTypeValueSet;
  }
  /**
   * Function to perform basic model validation (e.g., check if required elements are present).
   */
  public override doModelValidation():[string,string][] {
    var results:[string,string][] = super.doModelValidation();
    if (!this["resourceType"]) { results.push(["resourceType",'Missing required element: MolecularSequence.resourceType']); }
    if (this["identifier"]) { this.identifier.forEach((x) => { results.push(...x.doModelValidation()); }) }
    if (this["_type"]) { results.push(...this._type.doModelValidation()); }
    if (!this["coordinateSystem"]) { results.push(["coordinateSystem",'Missing required element: MolecularSequence.coordinateSystem']); }
    if (this["_coordinateSystem"]) { results.push(...this._coordinateSystem.doModelValidation()); }
    if (this["patient"]) { results.push(...this.patient.doModelValidation()); }
    if (this["specimen"]) { results.push(...this.specimen.doModelValidation()); }
    if (this["device"]) { results.push(...this.device.doModelValidation()); }
    if (this["performer"]) { results.push(...this.performer.doModelValidation()); }
    if (this["quantity"]) { results.push(...this.quantity.doModelValidation()); }
    if (this["referenceSeq"]) { results.push(...this.referenceSeq.doModelValidation()); }
    if (this["variant"]) { this.variant.forEach((x) => { results.push(...x.doModelValidation()); }) }
    if (this["_observedSeq"]) { results.push(...this._observedSeq.doModelValidation()); }
    if (this["quality"]) { this.quality.forEach((x) => { results.push(...x.doModelValidation()); }) }
    if (this["_readCoverage"]) { results.push(...this._readCoverage.doModelValidation()); }
    if (this["repository"]) { this.repository.forEach((x) => { results.push(...x.doModelValidation()); }) }
    if (this["pointer"]) { this.pointer.forEach((x) => { results.push(...x.doModelValidation()); }) }
    if (this["structureVariant"]) { this.structureVariant.forEach((x) => { results.push(...x.doModelValidation()); }) }
    return results;
  }
}
