// <auto-generated/>
// Contents of: hl7.fhir.r4.core version: 4.0.1
  // Primitive Naming Style: None
  // Complex Type / Resource Naming Style: PascalCase
  // Interaction Naming Style: None
  // Extension Support: NonPrimitive
// Minimum TypeScript Version: 3.7
import * as fhirModels from '../models'
/**
 * A sequence that is used as a reference to describe variants that are present in a sequence analyzed.
 */
export class MolecularSequenceReferenceSeq extends fhirModels.BackboneElement {
  /**
   * Structural unit composed of a nucleic acid molecule which controls its own replication through the interaction of specific proteins at one or more origins of replication ([SO:0000340](http://www.sequenceontology.org/browser/current_svn/term/SO:0000340)).
   */
  chromosome?: fhirModels.CodeableConcept;
  /**
   * The Genome Build used for reference, following GRCh build versions e.g. 'GRCh 37'.  Version number must be included if a versioned release of a primary build was used.
   */
  genomeBuild?: string;
  _genomeBuild?: fhirModels.Element;
  /**
   * A relative reference to a DNA strand based on gene orientation. The strand that contains the open reading frame of the gene is the "sense" strand, and the opposite complementary strand is the "antisense" strand.
   */
  orientation?: MolecularSequenceReferenceSeqOrientationEnum;
  _orientation?: fhirModels.Element;
  /**
   * Reference identifier of reference sequence submitted to NCBI. It must match the type in the MolecularSequence.type field. For example, the prefix, “NG_” identifies reference sequence for genes, “NM_” for messenger RNA transcripts, and “NP_” for amino acid sequences.
   */
  referenceSeqId?: fhirModels.CodeableConcept;
  /**
   * A pointer to another MolecularSequence entity as reference sequence.
   */
  referenceSeqPointer?: fhirModels.Reference;
  /**
   * A string like "ACGT".
   */
  referenceSeqString?: string;
  _referenceSeqString?: fhirModels.Element;
  /**
   * An absolute reference to a strand. The Watson strand is the strand whose 5'-end is on the short arm of the chromosome, and the Crick strand as the one whose 5'-end is on the long arm.
   */
  strand?: MolecularSequenceReferenceSeqStrandEnum;
  _strand?: fhirModels.Element;
  /**
   * End position of the window on the reference sequence. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  windowEnd?: number;
  _windowEnd?: fhirModels.Element;
  /**
   * Start position of the window on the reference sequence. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  windowStart?: number;
  _windowStart?: fhirModels.Element;
  /**
   * Default constructor
   */
  constructor(source: MolecularSequenceReferenceSeq) {
    super(source);
    if (source["chromosome"] !== undefined) { this.chromosome = source.chromosome; }
    if (source["genomeBuild"] !== undefined) { this.genomeBuild = source.genomeBuild; }
    if (source["_genomeBuild"] !== undefined) { this._genomeBuild = source._genomeBuild; }
    if (source["orientation"] !== undefined) { this.orientation = source.orientation; }
    if (source["_orientation"] !== undefined) { this._orientation = source._orientation; }
    if (source["referenceSeqId"] !== undefined) { this.referenceSeqId = source.referenceSeqId; }
    if (source["referenceSeqPointer"] !== undefined) { this.referenceSeqPointer = source.referenceSeqPointer; }
    if (source["referenceSeqString"] !== undefined) { this.referenceSeqString = source.referenceSeqString; }
    if (source["_referenceSeqString"] !== undefined) { this._referenceSeqString = source._referenceSeqString; }
    if (source["strand"] !== undefined) { this.strand = source.strand; }
    if (source["_strand"] !== undefined) { this._strand = source._strand; }
    if (source["windowEnd"] !== undefined) { this.windowEnd = source.windowEnd; }
    if (source["_windowEnd"] !== undefined) { this._windowEnd = source._windowEnd; }
    if (source["windowStart"] !== undefined) { this.windowStart = source.windowStart; }
    if (source["_windowStart"] !== undefined) { this._windowStart = source._windowStart; }
  }
}
/**
 * Code Values for the MolecularSequence.referenceSeq.orientation field
 */
export enum MolecularSequenceReferenceSeqOrientationEnum {
  SENSE = "sense",
  ANTISENSE = "antisense",
}
/**
 * Code Values for the MolecularSequence.referenceSeq.strand field
 */
export enum MolecularSequenceReferenceSeqStrandEnum {
  WATSON = "watson",
  CRICK = "crick",
}
/**
 * The definition of variant here originates from Sequence ontology ([variant_of](http://www.sequenceontology.org/browser/current_svn/term/variant_of)). This element can represent amino acid or nucleic sequence change(including insertion,deletion,SNP,etc.)  It can represent some complex mutation or segment variation with the assist of CIGAR string.
 */
export class MolecularSequenceVariant extends fhirModels.BackboneElement {
  /**
   * Extended CIGAR string for aligning the sequence with reference bases. See detailed documentation [here](http://support.illumina.com/help/SequencingAnalysisWorkflow/Content/Vault/Informatics/Sequencing_Analysis/CASAVA/swSEQ_mCA_ExtendedCIGARFormat.htm).
   */
  cigar?: string;
  _cigar?: fhirModels.Element;
  /**
   * End position of the variant on the reference sequence. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  end?: number;
  _end?: fhirModels.Element;
  /**
   * An allele is one of a set of coexisting sequence variants of a gene ([SO:0001023](http://www.sequenceontology.org/browser/current_svn/term/SO:0001023)).  Nucleotide(s)/amino acids from start position of sequence to stop position of sequence on the positive (+) strand of the observed  sequence. When the sequence  type is DNA, it should be the sequence on the positive (+) strand. This will lay in the range between variant.start and variant.end.
   */
  observedAllele?: string;
  _observedAllele?: fhirModels.Element;
  /**
   * An allele is one of a set of coexisting sequence variants of a gene ([SO:0001023](http://www.sequenceontology.org/browser/current_svn/term/SO:0001023)). Nucleotide(s)/amino acids from start position of sequence to stop position of sequence on the positive (+) strand of the reference sequence. When the sequence  type is DNA, it should be the sequence on the positive (+) strand. This will lay in the range between variant.start and variant.end.
   */
  referenceAllele?: string;
  _referenceAllele?: fhirModels.Element;
  /**
   * Start position of the variant on the  reference sequence. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  start?: number;
  _start?: fhirModels.Element;
  /**
   * A pointer to an Observation containing variant information.
   */
  variantPointer?: fhirModels.Reference;
  /**
   * Default constructor
   */
  constructor(source: MolecularSequenceVariant) {
    super(source);
    if (source["cigar"] !== undefined) { this.cigar = source.cigar; }
    if (source["_cigar"] !== undefined) { this._cigar = source._cigar; }
    if (source["end"] !== undefined) { this.end = source.end; }
    if (source["_end"] !== undefined) { this._end = source._end; }
    if (source["observedAllele"] !== undefined) { this.observedAllele = source.observedAllele; }
    if (source["_observedAllele"] !== undefined) { this._observedAllele = source._observedAllele; }
    if (source["referenceAllele"] !== undefined) { this.referenceAllele = source.referenceAllele; }
    if (source["_referenceAllele"] !== undefined) { this._referenceAllele = source._referenceAllele; }
    if (source["start"] !== undefined) { this.start = source.start; }
    if (source["_start"] !== undefined) { this._start = source._start; }
    if (source["variantPointer"] !== undefined) { this.variantPointer = source.variantPointer; }
  }
}
/**
 * Receiver Operator Characteristic (ROC) Curve  to give sensitivity/specificity tradeoff.
 */
export class MolecularSequenceQualityRoc extends fhirModels.BackboneElement {
  /**
   * Calculated fScore if the GQ score threshold was set to "score" field value.
   */
  fMeasure?: number[];
  _fMeasure?: fhirModels.Element[];
  /**
   * The number of false negatives if the GQ score threshold was set to "score" field value.
   */
  numFN?: number[];
  _numFN?: fhirModels.Element[];
  /**
   * The number of false positives if the GQ score threshold was set to "score" field value.
   */
  numFP?: number[];
  _numFP?: fhirModels.Element[];
  /**
   * The number of true positives if the GQ score threshold was set to "score" field value.
   */
  numTP?: number[];
  _numTP?: fhirModels.Element[];
  /**
   * Calculated precision if the GQ score threshold was set to "score" field value.
   */
  precision?: number[];
  _precision?: fhirModels.Element[];
  /**
   * Invidual data point representing the GQ (genotype quality) score threshold.
   */
  score?: number[];
  _score?: fhirModels.Element[];
  /**
   * Calculated sensitivity if the GQ score threshold was set to "score" field value.
   */
  sensitivity?: number[];
  _sensitivity?: fhirModels.Element[];
  /**
   * Default constructor
   */
  constructor(source: MolecularSequenceQualityRoc) {
    super(source);
    if (source["fMeasure"] !== undefined) { this.fMeasure = source.fMeasure; }
    if (source["_fMeasure"] !== undefined) { this._fMeasure = source._fMeasure; }
    if (source["numFN"] !== undefined) { this.numFN = source.numFN; }
    if (source["_numFN"] !== undefined) { this._numFN = source._numFN; }
    if (source["numFP"] !== undefined) { this.numFP = source.numFP; }
    if (source["_numFP"] !== undefined) { this._numFP = source._numFP; }
    if (source["numTP"] !== undefined) { this.numTP = source.numTP; }
    if (source["_numTP"] !== undefined) { this._numTP = source._numTP; }
    if (source["precision"] !== undefined) { this.precision = source.precision; }
    if (source["_precision"] !== undefined) { this._precision = source._precision; }
    if (source["score"] !== undefined) { this.score = source.score; }
    if (source["_score"] !== undefined) { this._score = source._score; }
    if (source["sensitivity"] !== undefined) { this.sensitivity = source.sensitivity; }
    if (source["_sensitivity"] !== undefined) { this._sensitivity = source._sensitivity; }
  }
}
/**
 * An experimental feature attribute that defines the quality of the feature in a quantitative way, such as a phred quality score ([SO:0001686](http://www.sequenceontology.org/browser/current_svn/term/SO:0001686)).
 */
export class MolecularSequenceQuality extends fhirModels.BackboneElement {
  /**
   * End position of the sequence. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  end?: number;
  _end?: fhirModels.Element;
  /**
   * Harmonic mean of Recall and Precision, computed as: 2 * precision * recall / (precision + recall).
   */
  fScore?: number;
  _fScore?: fhirModels.Element;
  /**
   * The number of false positives where the non-REF alleles in the Truth and Query Call Sets match (i.e. cases where the truth is 1/1 and the query is 0/1 or similar).
   */
  gtFP?: number;
  _gtFP?: fhirModels.Element;
  /**
   * Which method is used to get sequence quality.
   */
  method?: fhirModels.CodeableConcept;
  /**
   * QUERY.TP / (QUERY.TP + QUERY.FP).
   */
  precision?: number;
  _precision?: fhirModels.Element;
  /**
   * False positives, i.e. the number of sites in the Query Call Set for which there is no path through the Truth Call Set that is consistent with this site. Sites with correct variant but incorrect genotype are counted here.
   */
  queryFP?: number;
  _queryFP?: fhirModels.Element;
  /**
   * True positives, from the perspective of the query data, i.e. the number of sites in the Query Call Set for which there are paths through the Truth Call Set that are consistent with all of the alleles at this site, and for which there is an accurate genotype call for the event.
   */
  queryTP?: number;
  _queryTP?: fhirModels.Element;
  /**
   * TRUTH.TP / (TRUTH.TP + TRUTH.FN).
   */
  recall?: number;
  _recall?: fhirModels.Element;
  /**
   * Receiver Operator Characteristic (ROC) Curve  to give sensitivity/specificity tradeoff.
   */
  roc?: fhirModels.MolecularSequenceQualityRoc;
  /**
   * The score of an experimentally derived feature such as a p-value ([SO:0001685](http://www.sequenceontology.org/browser/current_svn/term/SO:0001685)).
   */
  score?: fhirModels.Quantity;
  /**
   * Gold standard sequence used for comparing against.
   */
  standardSequence?: fhirModels.CodeableConcept;
  /**
   * Start position of the sequence. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  start?: number;
  _start?: fhirModels.Element;
  /**
   * False negatives, i.e. the number of sites in the Truth Call Set for which there is no path through the Query Call Set that is consistent with all of the alleles at this site, or sites for which there is an inaccurate genotype call for the event. Sites with correct variant but incorrect genotype are counted here.
   */
  truthFN?: number;
  _truthFN?: fhirModels.Element;
  /**
   * True positives, from the perspective of the truth data, i.e. the number of sites in the Truth Call Set for which there are paths through the Query Call Set that are consistent with all of the alleles at this site, and for which there is an accurate genotype call for the event.
   */
  truthTP?: number;
  _truthTP?: fhirModels.Element;
  /**
   * INDEL / SNP / Undefined variant.
   */
  type: MolecularSequenceQualityTypeEnum;
  _type?: fhirModels.Element;
  /**
   * Default constructor
   */
  constructor(source: MolecularSequenceQuality) {
    super(source);
    if (source["end"] !== undefined) { this.end = source.end; }
    if (source["_end"] !== undefined) { this._end = source._end; }
    if (source["fScore"] !== undefined) { this.fScore = source.fScore; }
    if (source["_fScore"] !== undefined) { this._fScore = source._fScore; }
    if (source["gtFP"] !== undefined) { this.gtFP = source.gtFP; }
    if (source["_gtFP"] !== undefined) { this._gtFP = source._gtFP; }
    if (source["method"] !== undefined) { this.method = source.method; }
    if (source["precision"] !== undefined) { this.precision = source.precision; }
    if (source["_precision"] !== undefined) { this._precision = source._precision; }
    if (source["queryFP"] !== undefined) { this.queryFP = source.queryFP; }
    if (source["_queryFP"] !== undefined) { this._queryFP = source._queryFP; }
    if (source["queryTP"] !== undefined) { this.queryTP = source.queryTP; }
    if (source["_queryTP"] !== undefined) { this._queryTP = source._queryTP; }
    if (source["recall"] !== undefined) { this.recall = source.recall; }
    if (source["_recall"] !== undefined) { this._recall = source._recall; }
    if (source["roc"] !== undefined) { this.roc = source.roc; }
    if (source["score"] !== undefined) { this.score = source.score; }
    if (source["standardSequence"] !== undefined) { this.standardSequence = source.standardSequence; }
    if (source["start"] !== undefined) { this.start = source.start; }
    if (source["_start"] !== undefined) { this._start = source._start; }
    if (source["truthFN"] !== undefined) { this.truthFN = source.truthFN; }
    if (source["_truthFN"] !== undefined) { this._truthFN = source._truthFN; }
    if (source["truthTP"] !== undefined) { this.truthTP = source.truthTP; }
    if (source["_truthTP"] !== undefined) { this._truthTP = source._truthTP; }
    if (source["type"] === undefined) { throw 'Missing required element type';}
    this.type = source.type;
    if (source["_type"] !== undefined) { this._type = source._type; }
  }
}
/**
 * Code Values for the MolecularSequence.quality.type field
 */
export enum MolecularSequenceQualityTypeEnum {
  INDEL = "indel",
  SNP = "snp",
  UNKNOWN = "unknown",
}
/**
 * Configurations of the external repository. The repository shall store target's observedSeq or records related with target's observedSeq.
 */
export class MolecularSequenceRepository extends fhirModels.BackboneElement {
  /**
   * Id of the variant in this external repository. The server will understand how to use this id to call for more info about datasets in external repository.
   */
  datasetId?: string;
  _datasetId?: fhirModels.Element;
  /**
   * URI of an external repository which contains further details about the genetics data.
   */
  name?: string;
  _name?: fhirModels.Element;
  /**
   * Id of the read in this external repository.
   */
  readsetId?: string;
  _readsetId?: fhirModels.Element;
  /**
   * Click and see / RESTful API / Need login to see / RESTful API with authentication / Other ways to see resource.
   */
  type: MolecularSequenceRepositoryTypeEnum;
  _type?: fhirModels.Element;
  /**
   * URI of an external repository which contains further details about the genetics data.
   */
  url?: string;
  _url?: fhirModels.Element;
  /**
   * Id of the variantset in this external repository. The server will understand how to use this id to call for more info about variantsets in external repository.
   */
  variantsetId?: string;
  _variantsetId?: fhirModels.Element;
  /**
   * Default constructor
   */
  constructor(source: MolecularSequenceRepository) {
    super(source);
    if (source["datasetId"] !== undefined) { this.datasetId = source.datasetId; }
    if (source["_datasetId"] !== undefined) { this._datasetId = source._datasetId; }
    if (source["name"] !== undefined) { this.name = source.name; }
    if (source["_name"] !== undefined) { this._name = source._name; }
    if (source["readsetId"] !== undefined) { this.readsetId = source.readsetId; }
    if (source["_readsetId"] !== undefined) { this._readsetId = source._readsetId; }
    if (source["type"] === undefined) { throw 'Missing required element type';}
    this.type = source.type;
    if (source["_type"] !== undefined) { this._type = source._type; }
    if (source["url"] !== undefined) { this.url = source.url; }
    if (source["_url"] !== undefined) { this._url = source._url; }
    if (source["variantsetId"] !== undefined) { this.variantsetId = source.variantsetId; }
    if (source["_variantsetId"] !== undefined) { this._variantsetId = source._variantsetId; }
  }
}
/**
 * Code Values for the MolecularSequence.repository.type field
 */
export enum MolecularSequenceRepositoryTypeEnum {
  DIRECTLINK = "directlink",
  OPENAPI = "openapi",
  LOGIN = "login",
  OAUTH = "oauth",
  OTHER = "other",
}
/**
 * Structural variant outer.
 */
export class MolecularSequenceStructureVariantOuter extends fhirModels.BackboneElement {
  /**
   * Structural variant outer end. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  end?: number;
  _end?: fhirModels.Element;
  /**
   * Structural variant outer start. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  start?: number;
  _start?: fhirModels.Element;
  /**
   * Default constructor
   */
  constructor(source: MolecularSequenceStructureVariantOuter) {
    super(source);
    if (source["end"] !== undefined) { this.end = source.end; }
    if (source["_end"] !== undefined) { this._end = source._end; }
    if (source["start"] !== undefined) { this.start = source.start; }
    if (source["_start"] !== undefined) { this._start = source._start; }
  }
}
/**
 * Structural variant inner.
 */
export class MolecularSequenceStructureVariantInner extends fhirModels.BackboneElement {
  /**
   * Structural variant inner end. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
   */
  end?: number;
  _end?: fhirModels.Element;
  /**
   * Structural variant inner start. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
   */
  start?: number;
  _start?: fhirModels.Element;
  /**
   * Default constructor
   */
  constructor(source: MolecularSequenceStructureVariantInner) {
    super(source);
    if (source["end"] !== undefined) { this.end = source.end; }
    if (source["_end"] !== undefined) { this._end = source._end; }
    if (source["start"] !== undefined) { this.start = source.start; }
    if (source["_start"] !== undefined) { this._start = source._start; }
  }
}
/**
 * Information about chromosome structure variation.
 */
export class MolecularSequenceStructureVariant extends fhirModels.BackboneElement {
  /**
   * Used to indicate if the outer and inner start-end values have the same meaning.
   */
  exact?: boolean;
  _exact?: fhirModels.Element;
  /**
   * Structural variant inner.
   */
  inner?: fhirModels.MolecularSequenceStructureVariantInner;
  /**
   * Length of the variant chromosome.
   */
  length?: number;
  _length?: fhirModels.Element;
  /**
   * Structural variant outer.
   */
  outer?: fhirModels.MolecularSequenceStructureVariantOuter;
  /**
   * Information about chromosome structure variation DNA change type.
   */
  variantType?: fhirModels.CodeableConcept;
  /**
   * Default constructor
   */
  constructor(source: MolecularSequenceStructureVariant) {
    super(source);
    if (source["exact"] !== undefined) { this.exact = source.exact; }
    if (source["_exact"] !== undefined) { this._exact = source._exact; }
    if (source["inner"] !== undefined) { this.inner = source.inner; }
    if (source["length"] !== undefined) { this.length = source.length; }
    if (source["_length"] !== undefined) { this._length = source._length; }
    if (source["outer"] !== undefined) { this.outer = source.outer; }
    if (source["variantType"] !== undefined) { this.variantType = source.variantType; }
  }
}
/**
 * Raw data describing a biological sequence.
 */
export class MolecularSequence extends fhirModels.DomainResource {
  /**
   * Resource Type Name
   */
  readonly resourceType: string = "MolecularSequence";
  /**
   * Whether the sequence is numbered starting at 0 (0-based numbering or coordinates, inclusive start, exclusive end) or starting at 1 (1-based numbering, inclusive start and inclusive end).
   */
  coordinateSystem: number;
  _coordinateSystem?: fhirModels.Element;
  /**
   * The method for sequencing, for example, chip information.
   */
  device?: fhirModels.Reference;
  /**
   * A unique identifier for this particular sequence instance. This is a FHIR-defined id.
   */
  identifier?: fhirModels.Identifier[];
  /**
   * Sequence that was observed. It is the result marked by referenceSeq along with variant records on referenceSeq. This shall start from referenceSeq.windowStart and end by referenceSeq.windowEnd.
   */
  observedSeq?: string;
  _observedSeq?: fhirModels.Element;
  /**
   * The patient whose sequencing results are described by this resource.
   */
  patient?: fhirModels.Reference;
  /**
   * The organization or lab that should be responsible for this result.
   */
  performer?: fhirModels.Reference;
  /**
   * Pointer to next atomic sequence which at most contains one variant.
   */
  pointer?: fhirModels.Reference[];
  /**
   * An experimental feature attribute that defines the quality of the feature in a quantitative way, such as a phred quality score ([SO:0001686](http://www.sequenceontology.org/browser/current_svn/term/SO:0001686)).
   */
  quality?: fhirModels.MolecularSequenceQuality[];
  /**
   * The number of copies of the sequence of interest. (RNASeq).
   */
  quantity?: fhirModels.Quantity;
  /**
   * Coverage (read depth or depth) is the average number of reads representing a given nucleotide in the reconstructed sequence.
   */
  readCoverage?: number;
  _readCoverage?: fhirModels.Element;
  /**
   * A sequence that is used as a reference to describe variants that are present in a sequence analyzed.
   */
  referenceSeq?: fhirModels.MolecularSequenceReferenceSeq;
  /**
   * Configurations of the external repository. The repository shall store target's observedSeq or records related with target's observedSeq.
   */
  repository?: fhirModels.MolecularSequenceRepository[];
  /**
   * Specimen used for sequencing.
   */
  specimen?: fhirModels.Reference;
  /**
   * Information about chromosome structure variation.
   */
  structureVariant?: fhirModels.MolecularSequenceStructureVariant[];
  /**
   * Amino Acid Sequence/ DNA Sequence / RNA Sequence.
   */
  type?: MolecularSequenceTypeEnum;
  _type?: fhirModels.Element;
  /**
   * The definition of variant here originates from Sequence ontology ([variant_of](http://www.sequenceontology.org/browser/current_svn/term/variant_of)). This element can represent amino acid or nucleic sequence change(including insertion,deletion,SNP,etc.)  It can represent some complex mutation or segment variation with the assist of CIGAR string.
   */
  variant?: fhirModels.MolecularSequenceVariant[];
  /**
   * Default constructor
   */
  constructor(source: MolecularSequence) {
    super(source);
    if ((source['resourceType'] !== "MolecularSequence") || (source['resourceType'] !== undefined)) { throw 'Invalid resourceType for a MolecularSequence'; }
    if (source["coordinateSystem"] === undefined) { throw 'Missing required element coordinateSystem';}
    this.coordinateSystem = source.coordinateSystem;
    if (source["_coordinateSystem"] !== undefined) { this._coordinateSystem = source._coordinateSystem; }
    if (source["device"] !== undefined) { this.device = source.device; }
    if (source["identifier"] !== undefined) { this.identifier = source.identifier; }
    if (source["observedSeq"] !== undefined) { this.observedSeq = source.observedSeq; }
    if (source["_observedSeq"] !== undefined) { this._observedSeq = source._observedSeq; }
    if (source["patient"] !== undefined) { this.patient = source.patient; }
    if (source["performer"] !== undefined) { this.performer = source.performer; }
    if (source["pointer"] !== undefined) { this.pointer = source.pointer; }
    if (source["quality"] !== undefined) { this.quality = source.quality; }
    if (source["quantity"] !== undefined) { this.quantity = source.quantity; }
    if (source["readCoverage"] !== undefined) { this.readCoverage = source.readCoverage; }
    if (source["_readCoverage"] !== undefined) { this._readCoverage = source._readCoverage; }
    if (source["referenceSeq"] !== undefined) { this.referenceSeq = source.referenceSeq; }
    if (source["repository"] !== undefined) { this.repository = source.repository; }
    if (source["specimen"] !== undefined) { this.specimen = source.specimen; }
    if (source["structureVariant"] !== undefined) { this.structureVariant = source.structureVariant; }
    if (source["type"] !== undefined) { this.type = source.type; }
    if (source["_type"] !== undefined) { this._type = source._type; }
    if (source["variant"] !== undefined) { this.variant = source.variant; }
  }
}
/**
 * Code Values for the MolecularSequence.type field
 */
export enum MolecularSequenceTypeEnum {
  AA = "aa",
  DNA = "dna",
  RNA = "rna",
}
